// gitea-react-toolkit/dist/core/gitea-api/http/http.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.parseError = exports.defaultErrorMessages = exports.del = exports.patch = exports.put = exports.post = exports.get = exports.checkIfServerOnline = exports.extendConfig = exports.ERROR_NETWORK_ERROR = exports.ERROR_NETWORK_DISCONNECTED = exports.ERROR_SERVER_DISCONNECT_ERROR = exports.ERROR_SERVER_UNREACHABLE = exports.apiPath = void 0;

var _localforage = _interopRequireDefault(require("localforage"));

var _axios = _interopRequireDefault(require("axios"));

var _axiosCacheAdapter = require("axios-cache-adapter");

var _authentication = require("../authentication");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var apiPath = 'api/v1';
exports.apiPath = apiPath;
var DEFAULT_MAX_AGE = 1000;
var SERVER_ONLINE_STATUS = 200;
var ERROR_SERVER_UNREACHABLE = 'ERR_SERVER_UNREACHABLE';
exports.ERROR_SERVER_UNREACHABLE = ERROR_SERVER_UNREACHABLE;
var ERROR_SERVER_DISCONNECT_ERROR = 'ERROR_SERVER_DISCONNECT_ERROR';
exports.ERROR_SERVER_DISCONNECT_ERROR = ERROR_SERVER_DISCONNECT_ERROR;
var ERROR_NETWORK_DISCONNECTED = 'ERR_NETWORK_DISCONNECTED';
exports.ERROR_NETWORK_DISCONNECTED = ERROR_NETWORK_DISCONNECTED;
var ERROR_NETWORK_ERROR = 'Network Error';
exports.ERROR_NETWORK_ERROR = ERROR_NETWORK_ERROR;

var cacheStore = _localforage.default.createInstance({
  driver: [_localforage.default.INDEXEDDB],
  name: 'web-cache'
});

var api = (0, _axiosCacheAdapter.setup)({
  crossDomain: true,
  cache: {
    store: cacheStore,
    maxAge: 1 * 1 * 1 * 1 * DEFAULT_MAX_AGE,
    exclude: {
      query: false
    },
    key: function key(req) {
      var serialized = req.params instanceof URLSearchParams ? req.params.toString() : JSON.stringify(req.params) || '';
      return req.url + serialized;
    }
  }
});

var extendConfig = function extendConfig(config) {
  var headers = _objectSpread({}, config.headers);

  if (config && config.token) {
    var authHeaders = (0, _authentication.authorizationHeaders)({
      token: config.token
    });
    headers = _objectSpread(_objectSpread({}, config.headers), authHeaders);
  }

  var _config = _objectSpread(_objectSpread({
    baseURL: config.server
  }, config), {}, {
    headers: headers
  });

  return _config;
};

exports.extendConfig = extendConfig;

/**
 * create error object from errorMessage and flag that it was from checking network connection status.  Also add http
 *      response if given
 * @param {string} errorMessage
 * @param {object} response - http response
 */
function getServerError(errorMessage, response) {
  var error = new Error(errorMessage);
  error[ERROR_SERVER_DISCONNECT_ERROR] = true; // flag that this error was from checking if server was online

  if (response) {
    // if we have response, add it
    error['response'] = response;
  }

  return error;
}
/**
 * Make sure that we are still connected to the server.  First checks that we are connected to local network.  If not,
 *      it throws an exception.  If local network is connected it tries to verify that server is up.  If server is
 *      not up, it throws an exception.
 *    Note - when axios returns exception, it adds the response to error.  And we add flag
 *      ERROR_SERVER_DISCONNECT_ERROR to error to simplify determining that it was an error checking that
 *      server was online.
 *
 * @param {string} serverUrl - base path for server (e.g. 'https://git.door43.org')
 * @param {ExtendConfig} config - optional axios compatible config parameters
 */


var checkIfServerOnline = /*#__PURE__*/function () {
  var _ref = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee(serverUrl) {
    var _response;

    var config,
        response,
        errorMessage,
        serverIsResponding,
        _args = arguments;
    return regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            config = _args.length > 1 && _args[1] !== undefined ? _args[1] : {};

            if (navigator.onLine) {
              _context.next = 3;
              break;
            }

            throw getServerError(ERROR_NETWORK_DISCONNECTED, null);

          case 3:
            _context.prev = 3;
            _context.next = 6;
            return _axios.default.get("".concat(serverUrl, "/").concat(apiPath, "/version"), config);

          case 6:
            response = _context.sent;
            _context.next = 18;
            break;

          case 9:
            _context.prev = 9;
            _context.t0 = _context["catch"](3);
            errorMessage = _context.t0 && _context.t0.message ? _context.t0.message : '';

            if (!errorMessage.match(/network error/ig)) {
              _context.next = 16;
              break;
            }

            throw getServerError(ERROR_SERVER_UNREACHABLE, _context.t0 === null || _context.t0 === void 0 ? void 0 : _context.t0.response);

          case 16:
            _context.t0[ERROR_SERVER_DISCONNECT_ERROR] = true; // flag that this error was while checking if server was online

            throw _context.t0;

          case 18:
            serverIsResponding = ((_response = response) === null || _response === void 0 ? void 0 : _response.status) === SERVER_ONLINE_STATUS;

            if (serverIsResponding) {
              _context.next = 21;
              break;
            }

            throw getServerError(ERROR_SERVER_UNREACHABLE, response);

          case 21:
          case "end":
            return _context.stop();
        }
      }
    }, _callee, null, [[3, 9]]);
  }));

  return function checkIfServerOnline(_x) {
    return _ref.apply(this, arguments);
  };
}();
/**
 * do http get
 * @param {string} url
 * @param {object} params
 * @param {APIConfig|ExtendConfig} config - axios compatible config parameters
 * @param {boolean} [noCache] optional flag to disable caching
 * @param {boolean} [fullResponse] optional flag to return full http response including data and statusCode, useful if you want specifics such as http codes
 */


exports.checkIfServerOnline = checkIfServerOnline;

var get = /*#__PURE__*/function () {
  var _ref3 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee2(_ref2) {
    var url, params, config, noCache, fullResponse, _config, response, _params, data;

    return regeneratorRuntime.wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            url = _ref2.url, params = _ref2.params, config = _ref2.config, noCache = _ref2.noCache, fullResponse = _ref2.fullResponse;
            _config = config ? extendConfig(config) : {};
            _context2.prev = 2;

            if (!(noCache || _config.noCache)) {
              _context2.next = 10;
              break;
            }

            // also check config for noCache
            _params = _objectSpread({
              noCache: Math.random()
            }, params);
            _context2.next = 7;
            return _axios.default.get(url, _objectSpread(_objectSpread({}, _config), {}, {
              params: _params
            }));

          case 7:
            response = _context2.sent;
            _context2.next = 13;
            break;

          case 10:
            _context2.next = 12;
            return api.get(url, _objectSpread(_objectSpread({}, _config), {}, {
              params: params
            }));

          case 12:
            response = _context2.sent;

          case 13:
            _context2.next = 21;
            break;

          case 15:
            _context2.prev = 15;
            _context2.t0 = _context2["catch"](2);

            if (config.skipNetworkCheck) {
              _context2.next = 20;
              break;
            }

            _context2.next = 20;
            return checkIfServerOnline(config.server, config);

          case 20:
            // will arrive here if server is online
            if (fullResponse) {
              if (_context2.t0 !== null && _context2.t0 !== void 0 && _context2.t0.response) {
                // if http error, get response
                response = _context2.t0 === null || _context2.t0 === void 0 ? void 0 : _context2.t0.response;
              } else {
                // this is not http error, so get what we can from exception
                response = {
                  statusText: _context2.t0 === null || _context2.t0 === void 0 ? void 0 : _context2.t0.toString(),
                  status: 1
                };
              }
            }

          case 21:
            if (!fullResponse) {
              _context2.next = 23;
              break;
            }

            return _context2.abrupt("return", response);

          case 23:
            data = response ? response.data : null;
            return _context2.abrupt("return", data);

          case 25:
          case "end":
            return _context2.stop();
        }
      }
    }, _callee2, null, [[2, 15]]);
  }));

  return function get(_x2) {
    return _ref3.apply(this, arguments);
  };
}();

exports.get = get;

var post = /*#__PURE__*/function () {
  var _ref5 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee3(_ref4) {
    var url, payload, config, _config, _yield$axios$post, data;

    return regeneratorRuntime.wrap(function _callee3$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            url = _ref4.url, payload = _ref4.payload, config = _ref4.config;
            _config = extendConfig(config);
            _context3.next = 4;
            return _axios.default.post(url, payload, _config);

          case 4:
            _yield$axios$post = _context3.sent;
            data = _yield$axios$post.data;
            return _context3.abrupt("return", data);

          case 7:
          case "end":
            return _context3.stop();
        }
      }
    }, _callee3);
  }));

  return function post(_x3) {
    return _ref5.apply(this, arguments);
  };
}();

exports.post = post;

var put = /*#__PURE__*/function () {
  var _ref7 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee4(_ref6) {
    var url, payload, config, _config, _yield$axios$put, data;

    return regeneratorRuntime.wrap(function _callee4$(_context4) {
      while (1) {
        switch (_context4.prev = _context4.next) {
          case 0:
            url = _ref6.url, payload = _ref6.payload, config = _ref6.config;
            _config = extendConfig(config);
            _context4.next = 4;
            return _axios.default.put(url, payload, _config);

          case 4:
            _yield$axios$put = _context4.sent;
            data = _yield$axios$put.data;
            return _context4.abrupt("return", data);

          case 7:
          case "end":
            return _context4.stop();
        }
      }
    }, _callee4);
  }));

  return function put(_x4) {
    return _ref7.apply(this, arguments);
  };
}();

exports.put = put;

var patch = /*#__PURE__*/function () {
  var _ref9 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee5(_ref8) {
    var url, payload, config, _config, _yield$axios$patch, data;

    return regeneratorRuntime.wrap(function _callee5$(_context5) {
      while (1) {
        switch (_context5.prev = _context5.next) {
          case 0:
            url = _ref8.url, payload = _ref8.payload, config = _ref8.config;
            _config = extendConfig(config);
            _context5.next = 4;
            return _axios.default.patch(url, payload, _config);

          case 4:
            _yield$axios$patch = _context5.sent;
            data = _yield$axios$patch.data;
            return _context5.abrupt("return", data);

          case 7:
          case "end":
            return _context5.stop();
        }
      }
    }, _callee5);
  }));

  return function patch(_x5) {
    return _ref9.apply(this, arguments);
  };
}();

exports.patch = patch;

var del = /*#__PURE__*/function () {
  var _ref11 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee6(_ref10) {
    var url, payload, config, _config, _yield$axios$delete, data;

    return regeneratorRuntime.wrap(function _callee6$(_context6) {
      while (1) {
        switch (_context6.prev = _context6.next) {
          case 0:
            url = _ref10.url, payload = _ref10.payload, config = _ref10.config;
            _config = extendConfig(config);
            _config.data = payload;
            _context6.next = 5;
            return _axios.default.delete(url, _config);

          case 5:
            _yield$axios$delete = _context6.sent;
            data = _yield$axios$delete.data;
            return _context6.abrupt("return", data);

          case 8:
          case "end":
            return _context6.stop();
        }
      }
    }, _callee6);
  }));

  return function del(_x6) {
    return _ref11.apply(this, arguments);
  };
}();

exports.del = del;
var defaultErrorMessages = {
  actionText: 'Login',
  genericError: 'Something went wrong, please try again.',
  usernameError: 'Username does not exist.',
  passwordError: 'Password is invalid.',
  networkError: 'There is an issue with your network connection. Please try again.',
  serverError: 'There is an issue with the server please try again.'
};
exports.defaultErrorMessages = defaultErrorMessages;

var parseError = function parseError(_ref12) {
  var error = _ref12.error,
      _ref12$messages = _ref12.messages,
      messages = _ref12$messages === void 0 ? defaultErrorMessages : _ref12$messages;
  var errorMessage = error && error.message ? error.message : '';
  var friendlyError = {};

  if (errorMessage.match(ERROR_SERVER_UNREACHABLE)) {
    friendlyError = {
      errorMessage: messages.serverError,
      isRecoverable: false
    };
  } else if (errorMessage.match(ERROR_NETWORK_DISCONNECTED) || errorMessage.match(ERROR_NETWORK_ERROR)) {
    friendlyError = {
      errorMessage: messages.networkError,
      isRecoverable: false
    };
  } else {
    friendlyError = {
      errorMessage: messages.genericError,
      isRecoverable: true
    };
  }

  return friendlyError;
};

exports.parseError = parseError;
