// uw-content-validation/dist/core/books/books.js
"use strict";Object.defineProperty(exports, "__esModule", { value: true });exports.getEnglishBookName = getEnglishBookName;exports.isGoodEnglishBookName = isGoodEnglishBookName;exports.isOftenMissing = isOftenMissing;exports.testament = exports.versesInChapter = exports.isOneChapterBook = exports.chaptersInBook = exports.expectedVersesPerChapterList = exports.usfmNumberName = exports.isExtraBookID = exports.isOptionalValidBookID = exports.isValidBookID = exports.BibleBookData = void 0;var BibleBookData = {
  'GEN': { 'title': 'Genesis', 'usfm': '01-GEN', 'testament': 'old', 'verseCount': 1533, 'verseList': [31, 25, 24, 26, 32, 22, 24, 22, 29, 32, 32, 20, 18, 24, 21, 16, 27, 33, 38, 18, 34, 24, 20, 67, 34, 35, 46, 22, 35, 43, 55, 32, 20, 31, 29, 43, 36, 30, 23, 23, 57, 38, 34, 34, 28, 34, 31, 22, 33, 26] },
  'EXO': { 'title': 'Exodus', 'usfm': '02-EXO', 'testament': 'old', 'verseCount': 1213, 'verseList': [22, 25, 22, 31, 23, 30, 25, 32, 35, 29, 10, 51, 22, 31, 27, 36, 16, 27, 25, 26, 36, 31, 33, 18, 40, 37, 21, 43, 46, 38, 18, 35, 23, 35, 35, 38, 29, 31, 43, 38] },
  'LEV': { 'title': 'Leviticus', 'usfm': '03-LEV', 'testament': 'old', 'verseCount': 859, 'verseList': [17, 16, 17, 35, 19, 30, 38, 36, 24, 20, 47, 8, 59, 57, 33, 34, 16, 30, 37, 27, 24, 33, 44, 23, 55, 46, 34] },
  'NUM': { 'title': 'Numbers', 'usfm': '04-NUM', 'testament': 'old', 'verseCount': 1288, 'verseList': [54, 34, 51, 49, 31, 27, 89, 26, 23, 36, 35, 16, 33, 45, 41, 50, 13, 32, 22, 29, 35, 41, 30, 25, 18, 65, 23, 31, 40, 16, 54, 42, 56, 29, 34, 13] },
  'DEU': { 'title': 'Deuteronomy', 'usfm': '05-DEU', 'testament': 'old', 'verseCount': 959, 'verseList': [46, 37, 29, 49, 33, 25, 26, 20, 29, 22, 32, 32, 18, 29, 23, 22, 20, 22, 21, 20, 23, 30, 25, 22, 19, 19, 26, 68, 29, 20, 30, 52, 29, 12] },
  'JOS': { 'title': 'Joshua', 'usfm': '06-JOS', 'testament': 'old', 'verseCount': 658, 'verseList': [18, 24, 17, 24, 15, 27, 26, 35, 27, 43, 23, 24, 33, 15, 63, 10, 18, 28, 51, 9, 45, 34, 16, 33] },
  'JDG': { 'title': 'Judges', 'usfm': '07-JDG', 'testament': 'old', 'verseCount': 618, 'verseList': [36, 23, 31, 24, 31, 40, 25, 35, 57, 18, 40, 15, 25, 20, 20, 31, 13, 31, 30, 48, 25] },
  'RUT': { 'title': 'Ruth', 'usfm': '08-RUT', 'testament': 'old', 'verseCount': 85, 'verseList': [22, 23, 18, 22] },
  '1SA': { 'title': '1 Samuel', 'usfm': '09-1SA', 'testament': 'old', 'verseCount': 810, 'verseList': [28, 36, 21, 22, 12, 21, 17, 22, 27, 27, 15, 25, 23, 52, 35, 23, 58, 30, 24, 42, 15, 23, 29, 22, 44, 25, 12, 25, 11, 31, 13] },
  '2SA': { 'title': '2 Samuel', 'usfm': '10-2SA', 'testament': 'old', 'verseCount': 695, 'verseList': [27, 32, 39, 12, 25, 23, 29, 18, 13, 19, 27, 31, 39, 33, 37, 23, 29, 33, 43, 26, 22, 51, 39, 25] },
  '1KI': { 'title': '1 Kings', 'usfm': '11-1KI', 'testament': 'old', 'verseCount': 816, 'verseList': [53, 46, 28, 34, 18, 38, 51, 66, 28, 29, 43, 33, 34, 31, 34, 34, 24, 46, 21, 43, 29, 53] },
  '2KI': { 'title': '2 Kings', 'usfm': '12-2KI', 'testament': 'old', 'verseCount': 719, 'verseList': [18, 25, 27, 44, 27, 33, 20, 29, 37, 36, 21, 21, 25, 29, 38, 20, 41, 37, 37, 21, 26, 20, 37, 20, 30] },
  '1CH': { 'title': '1 Chronicles', 'usfm': '13-1CH', 'testament': 'old', 'verseCount': 942, 'verseList': [54, 55, 24, 43, 26, 81, 40, 40, 44, 14, 47, 40, 14, 17, 29, 43, 27, 17, 19, 8, 30, 19, 32, 31, 31, 32, 34, 21, 30] },
  '2CH': { 'title': '2 Chronicles', 'usfm': '14-2CH', 'testament': 'old', 'verseCount': 822, 'verseList': [17, 18, 17, 22, 14, 42, 22, 18, 31, 19, 23, 16, 22, 15, 19, 14, 19, 34, 11, 37, 20, 12, 21, 27, 28, 23, 9, 27, 36, 27, 21, 33, 25, 33, 27, 23] },
  'EZR': { 'title': 'Ezra', 'usfm': '15-EZR', 'testament': 'old', 'verseCount': 280, 'verseList': [11, 70, 13, 24, 17, 22, 28, 36, 15, 44] },
  'NEH': { 'title': 'Nehemiah', 'usfm': '16-NEH', 'testament': 'old', 'verseCount': 406, 'verseList': [11, 20, 32, 23, 19, 19, 73, 18, 38, 39, 36, 47, 31] },
  'EST': { 'title': 'Esther', 'usfm': '17-EST', 'testament': 'old', 'verseCount': 167, 'verseList': [22, 23, 15, 17, 14, 14, 10, 17, 32, 3] },
  'JOB': { 'title': 'Job', 'usfm': '18-JOB', 'testament': 'old', 'verseCount': 1070, 'verseList': [22, 13, 26, 21, 27, 30, 21, 22, 35, 22, 20, 25, 28, 22, 35, 22, 16, 21, 29, 29, 34, 30, 17, 25, 6, 14, 23, 28, 25, 31, 40, 22, 33, 37, 16, 33, 24, 41, 30, 24, 34, 17] },
  'PSA': { 'title': 'Psalms', 'usfm': '19-PSA', 'testament': 'old', 'verseCount': 2461, 'verseList': [6, 12, 8, 8, 12, 10, 17, 9, 20, 18, 7, 8, 6, 7, 5, 11, 15, 50, 14, 9, 13, 31, 6, 10, 22, 12, 14, 9, 11, 12, 24, 11, 22, 22, 28, 12, 40, 22, 13, 17, 13, 11, 5, 26, 17, 11, 9, 14, 20, 23, 19, 9, 6, 7, 23, 13, 11, 11, 17, 12, 8, 12, 11, 10, 13, 20, 7, 35, 36, 5, 24, 20, 28, 23, 10, 12, 20, 72, 13, 19, 16, 8, 18, 12, 13, 17, 7, 18, 52, 17, 16, 15, 5, 23, 11, 13, 12, 9, 9, 5, 8, 28, 22, 35, 45, 48, 43, 13, 31, 7, 10, 10, 9, 8, 18, 19, 2, 29, 176, 7, 8, 9, 4, 8, 5, 6, 5, 6, 8, 8, 3, 18, 3, 3, 21, 26, 9, 8, 24, 13, 10, 7, 12, 15, 21, 10, 20, 14, 9, 6] },
  'PRO': { 'title': 'Proverbs', 'usfm': '20-PRO', 'testament': 'old', 'verseCount': 915, 'verseList': [33, 22, 35, 27, 23, 35, 27, 36, 18, 32, 31, 28, 25, 35, 33, 33, 28, 24, 29, 30, 31, 29, 35, 34, 28, 28, 27, 28, 27, 33, 31] },
  'ECC': { 'title': 'Ecclesiastes', 'usfm': '21-ECC', 'testament': 'old', 'verseCount': 222, 'verseList': [18, 26, 22, 16, 20, 12, 29, 17, 18, 20, 10, 14] },
  'SNG': { 'title': 'Song of Songs', 'usfm': '22-SNG', 'testament': 'old', 'verseCount': 117, 'verseList': [17, 17, 11, 16, 16, 13, 13, 14] },
  'ISA': { 'title': 'Isaiah', 'usfm': '23-ISA', 'testament': 'old', 'verseCount': 1292, 'verseList': [31, 22, 26, 6, 30, 13, 25, 22, 21, 34, 16, 6, 22, 32, 9, 14, 14, 7, 25, 6, 17, 25, 18, 23, 12, 21, 13, 29, 24, 33, 9, 20, 24, 17, 10, 22, 38, 22, 8, 31, 29, 25, 28, 28, 25, 13, 15, 22, 26, 11, 23, 15, 12, 17, 13, 12, 21, 14, 21, 22, 11, 12, 19, 12, 25, 24] },
  'JER': { 'title': 'Jeremiah', 'usfm': '24-JER', 'testament': 'old', 'verseCount': 1364, 'verseList': [19, 37, 25, 31, 31, 30, 34, 22, 26, 25, 23, 17, 27, 22, 21, 21, 27, 23, 15, 18, 14, 30, 40, 10, 38, 24, 22, 17, 32, 24, 40, 44, 26, 22, 19, 32, 21, 28, 18, 16, 18, 22, 13, 30, 5, 28, 7, 47, 39, 46, 64, 34] },
  'LAM': { 'title': 'Lamentations', 'usfm': '25-LAM', 'testament': 'old', 'verseCount': 154, 'verseList': [22, 22, 66, 22, 22] },
  'EZK': { 'title': 'Ezekiel', 'usfm': '26-EZK', 'testament': 'old', 'verseCount': 1273, 'verseList': [28, 10, 27, 17, 17, 14, 27, 18, 11, 22, 25, 28, 23, 23, 8, 63, 24, 32, 14, 49, 32, 31, 49, 27, 17, 21, 36, 26, 21, 26, 18, 32, 33, 31, 15, 38, 28, 23, 29, 49, 26, 20, 27, 31, 25, 24, 23, 35] },
  'DAN': { 'title': 'Daniel', 'usfm': '27-DAN', 'testament': 'old', 'verseCount': 357, 'verseList': [21, 49, 30, 37, 31, 28, 28, 27, 27, 21, 45, 13] },
  'HOS': { 'title': 'Hosea', 'usfm': '28-HOS', 'testament': 'old', 'verseCount': 197, 'verseList': [11, 23, 5, 19, 15, 11, 16, 14, 17, 15, 12, 14, 16, 9] },
  'JOL': { 'title': 'Joel', 'usfm': '29-JOL', 'testament': 'old', 'verseCount': 73, 'verseList': [20, 32, 21] },
  'AMO': { 'title': 'Amos', 'usfm': '30-AMO', 'testament': 'old', 'verseCount': 146, 'verseList': [15, 16, 15, 13, 27, 14, 17, 14, 15] },
  'OBA': { 'title': 'Obadiah', 'usfm': '31-OBA', 'testament': 'old', 'verseCount': 21, 'verseList': [21] },
  'JON': { 'title': 'Jonah', 'usfm': '32-JON', 'testament': 'old', 'verseCount': 48, 'verseList': [17, 10, 10, 11] },
  'MIC': { 'title': 'Micah', 'usfm': '33-MIC', 'testament': 'old', 'verseCount': 105, 'verseList': [16, 13, 12, 13, 15, 16, 20] },
  'NAM': { 'title': 'Nahum', 'usfm': '34-NAM', 'testament': 'old', 'verseCount': 47, 'verseList': [15, 13, 19] },
  'HAB': { 'title': 'Habakkuk', 'usfm': '35-HAB', 'testament': 'old', 'verseCount': 56, 'verseList': [17, 20, 19] },
  'ZEP': { 'title': 'Zephaniah', 'usfm': '36-ZEP', 'testament': 'old', 'verseCount': 53, 'verseList': [18, 15, 20] },
  'HAG': { 'title': 'Haggai', 'usfm': '37-HAG', 'testament': 'old', 'verseCount': 38, 'verseList': [15, 23] },
  'ZEC': { 'title': 'Zechariah', 'usfm': '38-ZEC', 'testament': 'old', 'verseCount': 211, 'verseList': [21, 13, 10, 14, 11, 15, 14, 23, 17, 12, 17, 14, 9, 21] },
  'MAL': { 'title': 'Malachi', 'usfm': '39-MAL', 'testament': 'old', 'verseCount': 55, 'verseList': [14, 17, 18, 6] },
  'MAT': { 'title': 'Matthew', 'usfm': '41-MAT', 'testament': 'new', 'verseCount': 1071, 'verseList': [25, 23, 17, 25, 48, 34, 29, 34, 38, 42, 30, 50, 58, 36, 39, 28, 27, 35, 30, 34, 46, 46, 39, 51, 46, 75, 66, 20] },
  'MRK': { 'title': 'Mark', 'usfm': '42-MRK', 'testament': 'new', 'verseCount': 678, 'verseList': [45, 28, 35, 41, 43, 56, 37, 38, 50, 52, 33, 44, 37, 72, 47, 20] },
  'LUK': { 'title': 'Luke', 'usfm': '43-LUK', 'testament': 'new', 'verseCount': 1151, 'verseList': [80, 52, 38, 44, 39, 49, 50, 56, 62, 42, 54, 59, 35, 35, 32, 31, 37, 43, 48, 47, 38, 71, 56, 53] },
  'JHN': { 'title': 'John', 'usfm': '44-JHN', 'testament': 'new', 'verseCount': 879, 'verseList': [51, 25, 36, 54, 47, 71, 53, 59, 41, 42, 57, 50, 38, 31, 27, 33, 26, 40, 42, 31, 25] },
  'ACT': { 'title': 'Acts', 'usfm': '45-ACT', 'testament': 'new', 'verseCount': 1007, 'verseList': [26, 47, 26, 37, 42, 15, 60, 40, 43, 48, 30, 25, 52, 28, 41, 40, 34, 28, 41, 38, 40, 30, 35, 27, 27, 32, 44, 31] },
  'ROM': { 'title': 'Romans', 'usfm': '46-ROM', 'testament': 'new', 'verseCount': 433, 'verseList': [32, 29, 31, 25, 21, 23, 25, 39, 33, 21, 36, 21, 14, 23, 33, 27] },
  '1CO': { 'title': '1 Corinthians', 'usfm': '47-1CO', 'testament': 'new', 'verseCount': 437, 'verseList': [31, 16, 23, 21, 13, 20, 40, 13, 27, 33, 34, 31, 13, 40, 58, 24] },
  '2CO': { 'title': '2 Corinthians', 'usfm': '48-2CO', 'testament': 'new', 'verseCount': 257, 'verseList': [24, 17, 18, 18, 21, 18, 16, 24, 15, 18, 33, 21, 14] },
  'GAL': { 'title': 'Galatians', 'usfm': '49-GAL', 'testament': 'new', 'verseCount': 149, 'verseList': [24, 21, 29, 31, 26, 18] },
  'EPH': { 'title': 'Ephesians', 'usfm': '50-EPH', 'testament': 'new', 'verseCount': 155, 'verseList': [23, 22, 21, 32, 33, 24] },
  'PHP': { 'title': 'Philippians', 'usfm': '51-PHP', 'testament': 'new', 'verseCount': 104, 'verseList': [30, 30, 21, 23] },
  'COL': { 'title': 'Colossians', 'usfm': '52-COL', 'testament': 'new', 'verseCount': 95, 'verseList': [29, 23, 25, 18] },
  '1TH': { 'title': '1 Thessalonians', 'usfm': '53-1TH', 'testament': 'new', 'verseCount': 89, 'verseList': [10, 20, 13, 18, 28] },
  '2TH': { 'title': '2 Thessalonians', 'usfm': '54-2TH', 'testament': 'new', 'verseCount': 47, 'verseList': [12, 17, 18] },
  '1TI': { 'title': '1 Timothy', 'usfm': '55-1TI', 'testament': 'new', 'verseCount': 113, 'verseList': [20, 15, 16, 16, 25, 21] },
  '2TI': { 'title': '2 Timothy', 'usfm': '56-2TI', 'testament': 'new', 'verseCount': 83, 'verseList': [18, 26, 17, 22] },
  'TIT': { 'title': 'Titus', 'usfm': '57-TIT', 'testament': 'new', 'verseCount': 46, 'verseList': [16, 15, 15] },
  'PHM': { 'title': 'Philemon', 'usfm': '58-PHM', 'testament': 'new', 'verseCount': 25, 'verseList': [25] },
  'HEB': { 'title': 'Hebrews', 'usfm': '59-HEB', 'testament': 'new', 'verseCount': 303, 'verseList': [14, 18, 19, 16, 14, 20, 28, 13, 28, 39, 40, 29, 25] },
  'JAS': { 'title': 'James', 'usfm': '60-JAS', 'testament': 'new', 'verseCount': 108, 'verseList': [27, 26, 18, 17, 20] },
  '1PE': { 'title': '1 Peter', 'usfm': '61-1PE', 'testament': 'new', 'verseCount': 105, 'verseList': [25, 25, 22, 19, 14] },
  '2PE': { 'title': '2 Peter', 'usfm': '62-2PE', 'testament': 'new', 'verseCount': 61, 'verseList': [21, 22, 18] },
  '1JN': { 'title': '1 John', 'usfm': '63-1JN', 'testament': 'new', 'verseCount': 105, 'verseList': [10, 29, 24, 21, 21] },
  '2JN': { 'title': '2 John', 'usfm': '64-2JN', 'testament': 'new', 'verseCount': 13, 'verseList': [13] },
  '3JN': { 'title': '3 John', 'usfm': '65-3JN', 'testament': 'new', 'verseCount': 15, 'verseList': [15] },
  'JUD': { 'title': 'Jude', 'usfm': '66-JUD', 'testament': 'new', 'verseCount': 25, 'verseList': [25] },
  'REV': { 'title': 'Revelation', 'usfm': '67-REV', 'testament': 'new', 'verseCount': 404, 'verseList': [20, 29, 22, 11, 14, 17, 17, 13, 21, 11, 19, 17, 18, 20, 8, 21, 18, 24, 21, 15, 27, 21] } };exports.BibleBookData = BibleBookData;


var oftenMissingBCVList = [
// See https://en.wikipedia.org/wiki/List_of_New_Testament_verses_not_included_in_modern_English_translations
['NEH', 7, 68], // ?
['MAT', 16, 3],
['MAT', 17, 21],
['MAT', 18, 11],
// ['MAT', 20, 16b],
['MAT', 23, 14],
// ['MRK', 6, 11b],
['MRK', 7, 16],
['MRK', 9, 44], ['MRK', 9, 46],
['MRK', 11, 26],
['MRK', 15, 28],
['MRK', 16, 9], ['MRK', 16, 10], ['MRK', 16, 11], ['MRK', 16, 12], ['MRK', 16, 13], ['MRK', 16, 14],
['MRK', 16, 15], ['MRK', 16, 16], ['MRK', 16, 17], ['MRK', 16, 18], ['MRK', 16, 19], ['MRK', 16, 20],
// ['LUK', 4, 8b],
// ['LUK', 9, 55], ['LUK', 9, 56],
['LUK', 17, 36],
['LUK', 22, 43], // ?
['LUK', 22, 44], // ?
['LUK', 23, 17],
['JHN', 5, 3], ['JHN', 5, 4],
['JHN', 7, 53], // ?
['JHN', 8, 1], // ?
['ACT', 8, 37],
// ['ACT', 9, 5], ['ACT', 9, 6],
// ['ACT', 13, 42],
['ACT', 15, 34],
// ['ACT', 23, 9b],
['ACT', 24, 6], ['ACT', 24, 7], ['ACT', 24, 8],
['ACT', 28, 29],
['ROM', 16, 24],
['2CO', 13, 14], // ?
['1JN', 5, 7], ['1JN', 5, 8]];



var extraBookList = ['FRT', 'BAK'];
var isValidBookID = function isValidBookID(bookId) {
  return bookId.toUpperCase() in BibleBookData || extraBookList.includes(bookId);
};exports.isValidBookID = isValidBookID;
var isOptionalValidBookID = function isOptionalValidBookID(bookId) {
  return !bookId || bookId.toUpperCase() in BibleBookData || extraBookList.includes(bookId);
};exports.isOptionalValidBookID = isOptionalValidBookID;
var isExtraBookID = function isExtraBookID(bookId) {
  return extraBookList.includes(bookId);
};exports.isExtraBookID = isExtraBookID;


var usfmNumberName = function usfmNumberName(bookId) {
  try {return BibleBookData[bookId.toUpperCase()].usfm;}
  catch (err) {throw new Error("usfmNumberName() given invalid bookId: '".concat(bookId, "'"));}
};exports.usfmNumberName = usfmNumberName;

var expectedVersesPerChapterList = function expectedVersesPerChapterList(bookId) {
  // //parameterAssert(bookId.toUpperCase() !== 'OBS', `expectedVersesPerChapterList shouldn’t be passed '${bookId}'`);
  var verseList;
  try {
    verseList = BibleBookData[bookId.toUpperCase()].verseList;
  } catch (err) {
    throw new Error("expectedVersesPerChapterList() given invalid bookId: '".concat(bookId, "'"));
  }
  if (verseList === undefined) {
    throw new Error("expectedVersesPerChapterList(): Invalid bookId: '".concat(bookId, "'"));
  }
  return verseList;
};exports.expectedVersesPerChapterList = expectedVersesPerChapterList;

var chaptersInBook = function chaptersInBook(bookId) {
  return expectedVersesPerChapterList(bookId).length;
};exports.chaptersInBook = chaptersInBook;

var isOneChapterBook = function isOneChapterBook(bookId) {
  return chaptersInBook(bookId) === 1;
};exports.isOneChapterBook = isOneChapterBook;

var versesInChapter = function versesInChapter(bookId, chapter) {
  // //parameterAssert(bookId.toUpperCase() !== 'OBS', `versesInChapter shouldn’t be passed '${bookId}'`);
  var verses = expectedVersesPerChapterList(bookId)[chapter - 1];
  if (verses === undefined) {
    throw new Error("versesInChapter(".concat(bookId, ") given invalid chapter: ").concat(chapter));
  }
  return verses;
};exports.versesInChapter = versesInChapter;

var testament = function testament(bookId) {
  try {return BibleBookData[bookId.toUpperCase()].testament;}
  catch (err) {throw new Error("testament() given invalid bookId: '".concat(bookId, "'"));}
};exports.testament = testament;

function getEnglishBookName(bookId) {
  try {return BibleBookData[bookId.toUpperCase()].title;}
  catch (err) {throw new Error("getEnglishBookName() given invalid bookId: '".concat(bookId, "'"));}
}

function isGoodEnglishBookName(givenBookName) {
  // debugLog(`isGoodEnglishBookName(${givenBookName})…`);
  var partialMatches = [];
  var givenBookNameLower = givenBookName.toUpperCase();
  for (var bk in BibleBookData) {
    var thisBookName = BibleBookData[bk].title;
    // debugLog("thisBookName", thisBookName);
    if (thisBookName === givenBookName) return true;
    var thisBookNameLower = thisBookName.toUpperCase();
    if (thisBookNameLower === givenBookNameLower) return 1;
    if (thisBookNameLower.startsWith(givenBookNameLower)) partialMatches.push(thisBookName);
  }
  if (partialMatches.length === 1) return 2; // We got an unambiguous partial match, e.g., Gen for Genesis
  return false;
}

function isOftenMissing(bookID, C, V) {
  function matchBCV(entry) {return entry[0] === bookID && entry[1] === C && entry[2] === V;}

  return oftenMissingBCVList.find(matchBCV) !== undefined;
}
