// gitea-react-toolkit/dist/core/gitea-api/authentication.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.authenticate = exports.authorizationHeaders = exports.encodeAuthentication = void 0;

var _base = _interopRequireDefault(require("base-64"));

var _utf = _interopRequireDefault(require("utf8"));

var _users = require("./users");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

var encodeAuthentication = function encodeAuthentication(_ref) {
  var username = _ref.username,
      password = _ref.password,
      token = _ref.token;
  var authentication = '';

  if (token) {
    var sha1 = _typeof(token) === 'object' ? token.sha1 : token;
    authentication = "token ".concat(sha1);
  } else if (username && password) {
    var encoded = _base.default.encode(_utf.default.encode("".concat(username, ":").concat(password)));

    authentication = 'Basic ' + encoded;
  }

  return authentication;
};

exports.encodeAuthentication = encodeAuthentication;

var authorizationHeaders = function authorizationHeaders(_ref2) {
  var username = _ref2.username,
      password = _ref2.password,
      token = _ref2.token;
  var headers = {
    'Content-Type': '',
    'Authorization': ''
  };
  var authorization = encodeAuthentication({
    username: username,
    password: password,
    token: token
  });

  if (authorization) {
    headers = {
      'Content-Type': 'application/json',
      'Authorization': authorization
    };
  }

  return headers;
};

exports.authorizationHeaders = authorizationHeaders;
;

var authenticate = /*#__PURE__*/function () {
  var _ref4 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee(_ref3) {
    var username, password, config, token, user, _config, authHeaders, authentication;

    return regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            username = _ref3.username, password = _ref3.password, config = _ref3.config;
            _config = _objectSpread(_objectSpread({}, config), {}, {
              headers: {}
            });

            if (!(username && password)) {
              _context.next = 13;
              break;
            }

            authHeaders = authorizationHeaders({
              username: username,
              password: password
            });
            _config = _objectSpread(_objectSpread({}, config), {}, {
              headers: _objectSpread(_objectSpread({}, config.headers), authHeaders),
              noCache: true
            });
            _context.next = 7;
            return (0, _users.getUser)({
              username: username,
              config: _config
            });

          case 7:
            user = _context.sent;
            _context.next = 10;
            return (0, _users.ensureToken)({
              username: username,
              config: _config
            });

          case 10:
            token = _context.sent;
            authHeaders = authorizationHeaders({
              token: token
            });
            _config = _objectSpread(_objectSpread({}, config), {}, {
              headers: _objectSpread(_objectSpread({}, config.headers), authHeaders)
            });

          case 13:
            authentication = {
              user: user,
              token: token,
              config: _config
            };
            return _context.abrupt("return", authentication);

          case 15:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));

  return function authenticate(_x) {
    return _ref4.apply(this, arguments);
  };
}();

exports.authenticate = authenticate;
