// gitea-react-toolkit/dist/components/application-bar/ApplicationBar.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _core = require("@material-ui/core");

require("@material-ui/icons");

var _useStyles = require("./useStyles");

var _ = require("..");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function ApplicationBar(_ref) {
  var title = _ref.title,
      build = _ref.build,
      buttons = _ref.buttons,
      drawerMenu = _ref.drawerMenu,
      drawerMenuProps = _ref.drawerMenuProps,
      hideRepositoryMenu = _ref.hideRepositoryMenu;
  var classes = (0, _useStyles.useStyles)();

  var _useContext = (0, _react.useContext)(_.FileContext),
      file = _useContext.state;

  return /*#__PURE__*/_react.default.createElement("div", {
    className: classes.root
  }, /*#__PURE__*/_react.default.createElement(_core.AppBar, {
    position: "static",
    className: classes.appBar
  }, /*#__PURE__*/_react.default.createElement(_core.Toolbar, {
    "data-test": "application-bar"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: classes.menuButton
  }, /*#__PURE__*/_react.default.createElement(_.DrawerMenu, drawerMenuProps, drawerMenu)), /*#__PURE__*/_react.default.createElement(_core.Typography, {
    variant: "h6",
    color: "inherit",
    className: classes.grow,
    noWrap: true
  }, title, build && /*#__PURE__*/_react.default.createElement(_core.Typography, {
    variant: "caption",
    color: "inherit"
  }, " build ", build)), /*#__PURE__*/_react.default.createElement(_core.Typography, {
    variant: "subtitle2",
    color: "inherit",
    className: classes.grow,
    noWrap: true
  }, file ? file.filepath : ''), /*#__PURE__*/_react.default.createElement("div", {
    className: classes.grow
  }), buttons, !hideRepositoryMenu ? /*#__PURE__*/_react.default.createElement(_.RepositoryMenu, null) : null, /*#__PURE__*/_react.default.createElement(_.UserMenu, null))));
}

ApplicationBar.defaultProps = {
  drawerMenuProps: {},
  hideRepositoryMenu: false
};
ApplicationBar.propTypes = {
  /** The title string or jsx to be displayed. */
  title: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.element]),

  /** Additional buttons to be displayed. */
  buttons: _propTypes.default.element,

  /** Component to render inside of the drawer menu. */
  drawerMenu: _propTypes.default.element,

  /** Drawer menu props. */
  drawerMenuProps: _propTypes.default.object
};
var _default = ApplicationBar;
exports.default = _default;
