// gitea-react-toolkit/dist/core/gitea-api/repos/repos.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.deleteRepo = exports.updateRepo = exports.ensureRepo = exports.readRepo = exports.createRepo = void 0;

var _path = _interopRequireDefault(require("path"));

var _ = require("..");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

// POST /api/v1/user/repos
var createRepo = /*#__PURE__*/function () {
  var _ref2 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee(_ref) {
    var repo, settings, config, url, payload, response;
    return regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            repo = _ref.repo, settings = _ref.settings, config = _ref.config;
            url = _path.default.join(_.apiPath, 'user', 'repos');
            payload = _objectSpread({
              name: repo,
              auto_init: true
            }, settings);
            _context.next = 5;
            return (0, _.post)({
              url: url,
              payload: payload,
              config: config
            });

          case 5:
            response = _context.sent;
            return _context.abrupt("return", response);

          case 7:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));

  return function createRepo(_x) {
    return _ref2.apply(this, arguments);
  };
}();

exports.createRepo = createRepo;

// GET /api/v1/repos/{owner}/{repo}
var readRepo = /*#__PURE__*/function () {
  var _ref4 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee2(_ref3) {
    var owner, repo, config, url, response;
    return regeneratorRuntime.wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            owner = _ref3.owner, repo = _ref3.repo, config = _ref3.config;
            url = _path.default.join(_.apiPath, 'repos', owner, repo);
            _context2.prev = 2;
            _context2.next = 5;
            return (0, _.get)({
              url: url,
              config: config,
              noCache: true
            });

          case 5:
            response = _context2.sent;
            _context2.next = 11;
            break;

          case 8:
            _context2.prev = 8;
            _context2.t0 = _context2["catch"](2);
            response = null;

          case 11:
            return _context2.abrupt("return", response);

          case 12:
          case "end":
            return _context2.stop();
        }
      }
    }, _callee2, null, [[2, 8]]);
  }));

  return function readRepo(_x2) {
    return _ref4.apply(this, arguments);
  };
}();

exports.readRepo = readRepo;

var ensureRepo = /*#__PURE__*/function () {
  var _ref6 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee3(_ref5) {
    var owner, repo, settings, config, repository;
    return regeneratorRuntime.wrap(function _callee3$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            owner = _ref5.owner, repo = _ref5.repo, settings = _ref5.settings, config = _ref5.config;
            _context3.next = 3;
            return readRepo({
              owner: owner,
              repo: repo,
              config: config
            });

          case 3:
            repository = _context3.sent;

            if (repository) {
              _context3.next = 8;
              break;
            }

            _context3.next = 7;
            return createRepo({
              repo: repo,
              settings: settings,
              config: config
            });

          case 7:
            repository = _context3.sent;

          case 8:
            return _context3.abrupt("return", repository);

          case 9:
          case "end":
            return _context3.stop();
        }
      }
    }, _callee3);
  }));

  return function ensureRepo(_x3) {
    return _ref6.apply(this, arguments);
  };
}(); // PATCH /api/v1/repos/{owner}/{repo}


exports.ensureRepo = ensureRepo;

var updateRepo = /*#__PURE__*/function () {
  var _ref8 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee4(_ref7) {
    var owner, repo, settings, config, url, payload, response;
    return regeneratorRuntime.wrap(function _callee4$(_context4) {
      while (1) {
        switch (_context4.prev = _context4.next) {
          case 0:
            owner = _ref7.owner, repo = _ref7.repo, settings = _ref7.settings, config = _ref7.config;
            url = _path.default.join(_.apiPath, 'repos', owner, repo);
            payload = _objectSpread({}, settings);
            _context4.prev = 3;
            _context4.next = 6;
            return (0, _.patch)({
              url: url,
              payload: payload,
              config: config
            });

          case 6:
            response = _context4.sent;
            _context4.next = 12;
            break;

          case 9:
            _context4.prev = 9;
            _context4.t0 = _context4["catch"](3);
            response = null;

          case 12:
            return _context4.abrupt("return", response);

          case 13:
          case "end":
            return _context4.stop();
        }
      }
    }, _callee4, null, [[3, 9]]);
  }));

  return function updateRepo(_x4) {
    return _ref8.apply(this, arguments);
  };
}(); // DELETE /api/v1/repos/{owner}/{repo}


exports.updateRepo = updateRepo;

var deleteRepo = /*#__PURE__*/function () {
  var _ref10 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee5(_ref9) {
    var owner, repo, config, url, response;
    return regeneratorRuntime.wrap(function _callee5$(_context5) {
      while (1) {
        switch (_context5.prev = _context5.next) {
          case 0:
            owner = _ref9.owner, repo = _ref9.repo, config = _ref9.config;
            url = _path.default.join(_.apiPath, 'repos', owner, repo);
            _context5.prev = 2;
            _context5.next = 5;
            return (0, _.del)({
              url: url,
              config: config,
              payload: {}
            });

          case 5:
            response = _context5.sent;
            _context5.next = 11;
            break;

          case 8:
            _context5.prev = 8;
            _context5.t0 = _context5["catch"](2);
            response = null;

          case 11:
            return _context5.abrupt("return", response);

          case 12:
          case "end":
            return _context5.stop();
        }
      }
    }, _callee5, null, [[2, 8]]);
  }));

  return function deleteRepo(_x5) {
    return _ref10.apply(this, arguments);
  };
}();

exports.deleteRepo = deleteRepo;
