// gitea-react-toolkit/dist/core/gitea-api/repos/forks.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createFork = exports.readForks = void 0;

var _path = _interopRequireDefault(require("path"));

var _ = require("..");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

// GET /repos/{owner}/{repo}/forks
var readForks = /*#__PURE__*/function () {
  var _ref2 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee(_ref) {
    var owner, repo, config, url, response;
    return regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            owner = _ref.owner, repo = _ref.repo, config = _ref.config;
            url = _path.default.join(_.apiPath, 'repos', owner, repo, 'forks');
            _context.next = 4;
            return (0, _.get)({
              url: url,
              config: config
            });

          case 4:
            response = _context.sent;
            return _context.abrupt("return", response);

          case 6:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));

  return function readForks(_x) {
    return _ref2.apply(this, arguments);
  };
}();

exports.readForks = readForks;

// POST /repos/{owner}/{repo}/forks
var createFork = /*#__PURE__*/function () {
  var _ref4 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee2(_ref3) {
    var owner, repo, organization, config, url, payload, response;
    return regeneratorRuntime.wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            owner = _ref3.owner, repo = _ref3.repo, organization = _ref3.organization, config = _ref3.config;
            url = _path.default.join(_.apiPath, 'repos', owner, repo, 'forks');
            payload = organization ? {
              organization: organization
            } : {};
            _context2.next = 5;
            return (0, _.post)({
              url: url,
              payload: payload,
              config: config
            });

          case 5:
            response = _context2.sent;
            return _context2.abrupt("return", response);

          case 7:
          case "end":
            return _context2.stop();
        }
      }
    }, _callee2);
  }));

  return function createFork(_x2) {
    return _ref4.apply(this, arguments);
  };
}();

exports.createFork = createFork;
