// uw-languages-rcl/dist/core/useLanguages.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useLanguages = useLanguages;
exports.default = void 0;

var _concat = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/concat"));

var _regenerator = _interopRequireDefault(require("@babel/runtime-corejs3/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/asyncToGenerator"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/slicedToArray"));

var _react = require("react");

var _useDeepCompareEffect = _interopRequireDefault(require("use-deep-compare-effect"));

/* The languages JSON is an array of objects. Sample:
{
  "pk": 6,
  "ang": "Afar",
  "lr": "Africa",
  "gw": false,
  "hc": "ET",
  "ln": "Afaraf",
  "lc": "aa",
  "ld": "ltr",
  "alt": [
    "Afaraf",
    "Danakil",
    "Denkel",
    "Adal",
    "Afar Af",
    "Qafar",
    "Baadu (Ba'adu)"
  ],
  "cc": ["DJ", "ER", "ET", "US", "CA"]
},
*/
function useLanguages() {
  var _useState = (0, _react.useState)([]),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      languages = _useState2[0],
      setLanguages = _useState2[1];

  (0, _useDeepCompareEffect.default)(function () {
    function getLanguages() {
      return _getLanguages.apply(this, arguments);
    }

    function _getLanguages() {
      _getLanguages = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee() {
        var langs, _langs;

        return _regenerator.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return fetch('https://td.unfoldingword.org/exports/langnames.json');

              case 2:
                langs = _context.sent;
                _context.next = 5;
                return langs.json();

              case 5:
                _langs = _context.sent;
                setLanguages(_langs);

              case 7:
              case "end":
                return _context.stop();
            }
          }
        }, _callee);
      }));
      return _getLanguages.apply(this, arguments);
    }

    if (languages.length === 0) {
      //console.log("languages is empty... fetching")
      getLanguages();
    }
  }, [languages]);
  var formatLanguage = (0, _react.useCallback)(function (lg) {
    var _context2, _context3, _context4;

    // pattern for uw format: (am) Amharic – አማርኛ (Africa Gateway)
    var lc = lg.lc ? lg.lc : 'UNK';
    var ang = lg.ang ? lg.ang : 'UNK';
    var ln = lg.ln ? lg.ln : 'UNK';
    var lr = lg.lr ? lg.lr : 'UNK';
    return (0, _concat.default)(_context2 = (0, _concat.default)(_context3 = (0, _concat.default)(_context4 = "(".concat(lc, ") ")).call(_context4, ang, " - ")).call(_context3, ln, " (")).call(_context2, lr, ")");
  }, []);

  var getAllLanguages = function getAllLanguages() {
    return languages;
  };

  var getLanguage = (0, _react.useCallback)(function (languageId) {
    for (var i = 0; i < languages.length; i++) {
      if (languages[i].lc === languageId) {
        return languages[i];
      }
    }

    return {};
  }, [languages]);
  var getGatewayLanguages = (0, _react.useCallback)(function () {
    var _languages = [];

    for (var i = 0; i < languages.length; i++) {
      if (languages[i].gw) {
        _languages.push(formatLanguage(languages[i]));
      }
    }

    return _languages;
  }, [languages, formatLanguage]);
  var actions = {
    getAllLanguages: getAllLanguages,
    getLanguage: getLanguage,
    getGatewayLanguages: getGatewayLanguages,
    formatLanguage: formatLanguage
  }; // We return these computed values for the passed date prop to our hook

  return {
    state: languages,
    actions: actions
  };
}

;
var _default = useLanguages;
exports.default = _default;
