// gitea-react-toolkit/dist/components/application-bar/RepositoryMenu.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _core = require("@material-ui/core");

var _icons = require("@material-ui/icons");

var _useStyles = require("./useStyles");

var _ = require("..");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function RepositoryMenu() {
  var classes = (0, _useStyles.useStyles)();

  var _useState = (0, _react.useState)(false),
      _useState2 = _slicedToArray(_useState, 2),
      modal = _useState2[0],
      setModal = _useState2[1];

  var _ref = (0, _react.useContext)(_.RepositoryContext) || {},
      repository = _ref.state,
      actions = _ref.actions,
      repositoryComponent = _ref.component;

  var _ref2 = (0, _react.useContext)(_.FileContext) || {},
      file = _ref2.state,
      fileStateValues = _ref2.stateValues,
      fileActions = _ref2.actions;

  var _ref3 = repository || {},
      name = _ref3.name,
      avatar_url = _ref3.avatar_url,
      owner = _ref3.owner,
      full_name = _ref3.full_name;

  var _onDelete = (0, _react.useCallback)(function () {
    if (actions !== null && actions !== void 0 && actions.close) {
      if (fileActions !== null && fileActions !== void 0 && fileActions.onConfirmClose) {
        if (fileActions.onConfirmClose()) {
          actions.close();
        }
      } else {
        actions.close();
      }
    }
  }, [actions.close, file, fileActions === null || fileActions === void 0 ? void 0 : fileActions.onConfirmClose]);

  var handleOpen = (0, _react.useCallback)(function () {
    setModal(true);
  }, []);
  var button;

  if (name && owner) {
    var avatarComponent = /*#__PURE__*/_react.default.createElement(_core.Avatar, {
      src: avatar_url || owner.avatar_url
    });

    button = /*#__PURE__*/_react.default.createElement(_core.Chip, {
      id: "chip_".concat(name),
      "data-test": "repository-item-icon",
      "aria-label": "repository-item-icon",
      avatar: avatarComponent,
      label: /*#__PURE__*/_react.default.createElement("span", {
        "data-test": "repository-item-chip"
      }, name),
      onDelete: _onDelete,
      color: "primary"
    });
  } else {
    button = /*#__PURE__*/_react.default.createElement(_core.IconButton, {
      onClick: handleOpen,
      color: "inherit"
    }, /*#__PURE__*/_react.default.createElement(_icons.FolderShared, null));
  }

  ;

  var modalComponent = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null);

  if (modal && !full_name) {
    modalComponent = /*#__PURE__*/_react.default.createElement(_core.Modal, {
      "data-test": "repository-menu-modal",
      open: true,
      onClose: function onClose() {
        return setModal(false);
      }
    }, /*#__PURE__*/_react.default.createElement(_core.Paper, {
      className: classes.modal
    }, repositoryComponent));
  }

  ;
  return /*#__PURE__*/_react.default.createElement("div", {
    "data-test": "repository-menu"
  }, button, modalComponent);
}

;
RepositoryMenu.propTypes = {};
var _default = RepositoryMenu;
exports.default = _default;
