// gitea-react-toolkit/dist/components/organizations/Organizations.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _core = require("@material-ui/core");

var _ = require("../");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function Organizations(_ref) {
  var url = _ref.url,
      _ref$organization = _ref.organization,
      organization = _ref$organization === void 0 ? {} : _ref$organization,
      urls = _ref.urls,
      organizations = _ref.organizations,
      onOrganization = _ref.onOrganization,
      config = _ref.config,
      _ref$messages = _ref.messages;
  _ref$messages = _ref$messages === void 0 ? {} : _ref$messages;
  var _ref$messages$primary = _ref$messages.primaryError,
      primaryError = _ref$messages$primary === void 0 ? 'No organizations' : _ref$messages$primary,
      _ref$messages$seconda = _ref$messages.secondaryError,
      secondaryError = _ref$messages$seconda === void 0 ? 'Please provide valid organization objects or urls.' : _ref$messages$seconda;
  var updateOrganization = (0, _react.useCallback)(function (repo) {
    onOrganization(repo);
  }, [onOrganization]);
  var components = [];

  if (organizations && organizations.length > 0) {
    components = organizations.map(function (_organization) {
      return /*#__PURE__*/_react.default.createElement(_.Organization, {
        key: _organization.id,
        selected: organization && _organization.id === organization.id,
        organization: _organization,
        onOrganization: updateOrganization,
        config: config
      });
    });
  } else if (urls && urls.length > 0) {
    components = urls.map(function (_url, index) {
      return /*#__PURE__*/_react.default.createElement(_.Organization, {
        selected: _url === url,
        key: index,
        url: _url,
        onOrganization: updateOrganization,
        config: config,
        messages: {
          primaryError: primaryError,
          secondaryError: secondaryError
        }
      });
    });
  } else {
    components = [/*#__PURE__*/_react.default.createElement(_.Organization, {
      key: 0,
      organization: organization,
      config: config,
      messages: {
        primaryError: primaryError,
        secondaryError: secondaryError
      }
    })];
  }

  return /*#__PURE__*/_react.default.createElement(_core.List, null, components);
}

Organizations.propTypes = {
  /** Url to get organization data, if organization data is not provided. */
  url: _propTypes.default.string,

  /** Urls array to get organization data, if organization data is not provided. */
  urls: _propTypes.default.array,

  /** Organizations data array to render, if urls not provided. */
  organizations: _propTypes.default.array,

  /** Function to call when organization is selected. */
  onOrganization: _propTypes.default.func.isRequired,

  /** Configuration required if paths are provided as URL. */
  config: _propTypes.default.shape({
    server: _propTypes.default.string.isRequired
  }),

  /** The currently selected organization */
  organization: _propTypes.default.shape({
    avatar_url: _propTypes.default.string.isRequired,
    description: _propTypes.default.string.isRequired,
    full_name: _propTypes.default.string.isRequired,
    id: _propTypes.default.number.isRequired,
    location: _propTypes.default.string.isRequired,
    username: _propTypes.default.string.isRequired,
    visibility: _propTypes.default.string.isRequired,
    website: _propTypes.default.string.isRequired
  }),
  messages: _propTypes.default.shape({
    primaryError: _propTypes.default.string,
    secondaryError: _propTypes.default.string
  })
};
var _default = Organizations;
exports.default = _default;
