// gitea-react-toolkit/dist/components/tree-blob/Tree.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styles = require("@material-ui/core/styles");

var _core = require("@material-ui/core");

var _helpers = require("./helpers");

var _ = require(".");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var useStyles = (0, _styles.makeStyles)(function (theme) {
  return {
    list: {
      paddingTop: 0,
      paddingBottom: 0
    }
  };
});
/**
 * A Listing Component to render an array of Git Tree objects.
 */

function Tree(_ref) {
  var tree = _ref.tree,
      url = _ref.url,
      config = _ref.config,
      selected = _ref.selected,
      blob = _ref.blob,
      onBlob = _ref.onBlob,
      depth = _ref.depth,
      filepath = _ref.filepath,
      comparer = _ref.comparer;
  var classes = useStyles();

  var _useState = (0, _react.useState)(tree || []),
      _useState2 = _slicedToArray(_useState, 2),
      _tree = _useState2[0],
      setTree = _useState2[1];

  var _selectedPath;

  if (blob) {
    _selectedPath = blob.filepath.split('/')[depth - 2];
  }

  var _useState3 = (0, _react.useState)(_selectedPath),
      _useState4 = _slicedToArray(_useState3, 2),
      selectedPath = _useState4[0],
      setSelectedPath = _useState4[1];

  var updateTree = /*#__PURE__*/function () {
    var _ref2 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee() {
      var __tree;

      return regeneratorRuntime.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              _context.next = 2;
              return (0, _helpers.fetchTree)({
                url: url,
                config: config,
                comparer: comparer
              });

            case 2:
              __tree = _context.sent;
              setTree(__tree);

            case 4:
            case "end":
              return _context.stop();
          }
        }
      }, _callee);
    }));

    return function updateTree() {
      return _ref2.apply(this, arguments);
    };
  }();

  var emptyTree = !_tree || _tree.length === 0;

  if (selected && emptyTree) {
    updateTree();
  }

  var components = [];

  if (_tree) {
    components = _tree.map(function (object, index) {
      var _selected = object.path === selectedPath;

      var component;

      if (object.type === 'tree') {
        component = /*#__PURE__*/_react.default.createElement(_.TreeObject, _extends({}, object, {
          selected: _selected,
          depth: depth,
          onBlob: onBlob,
          filepath: filepath
        }));
      } else if (object.type === 'blob') {
        component = /*#__PURE__*/_react.default.createElement(_.BlobObject, {
          blob: object,
          onBlob: onBlob,
          selected: _selected,
          depth: depth,
          filepath: filepath
        });
      }

      var onSelectedPath = function onSelectedPath() {
        return setSelectedPath(object.path);
      };

      return /*#__PURE__*/_react.default.createElement("div", {
        key: index,
        onClick: onSelectedPath
      }, component);
    });
  }

  return /*#__PURE__*/_react.default.createElement(_core.Collapse, {
    in: selected,
    timeout: "auto",
    unmountOnExit: true
  }, /*#__PURE__*/_react.default.createElement(_core.List, {
    "data-test": "file-tree",
    dense: true,
    className: classes.list
  }, components));
}

Tree.propTypes = {
  /** An array of paths from the Gitea file tree api. */
  tree: _propTypes.default.arrayOf(_propTypes.default.shape({
    path: _propTypes.default.string.isRequired,
    type: _propTypes.default.oneOf(['tree', 'blob']).isRequired
  })),

  /** The Url to fetch the listing if listing is not provided. */
  url: _propTypes.default.string,

  /** If url is relative, pass the server in the config object. */
  config: _propTypes.default.shape({
    server: _propTypes.default.string
  }),

  /** Set if the Listing is currently selected, which will expand the collapsed view. */
  selected: _propTypes.default.bool,

  /** Function to propogate when the Blob is selected. */
  onBlob: _propTypes.default.func,

  /** The depth of the path in the tree sets the inset of the component. */
  depth: _propTypes.default.number,

  /** The nested filepath that will concatenate. */
  filepath: _propTypes.default.string,

  /** Blob data to render, if url not provided. */
  blob: _propTypes.default.shape({
    /** The filepath in the Git Tree Blob Object */
    path: _propTypes.default.string.isRequired,

    /** The url in the Git Tree Blob Object */
    url: _propTypes.default.string,

    /** The content size of the Git Tree Blob Object */
    size: _propTypes.default.number
  })
};
Tree.defaultProps = {
  selected: false,
  depth: 1
};
var _default = Tree;
exports.default = _default;
