// gitea-react-toolkit/dist/components/tree-blob/TreeObject.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _path = _interopRequireDefault(require("path"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styles = require("@material-ui/core/styles");

var _core = require("@material-ui/core");

var _icons = require("@material-ui/icons");

var _ = require("./");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var useStyles = (0, _styles.makeStyles)(function (theme) {
  return {
    root: {
      paddingRight: '0.7em'
    },
    pathText: {
      paddingLeft: '0.7em'
    }
  };
});
/**
 * A TreeObject Component to render a Git Tree tree object.
 */

function TreeObject(_ref) {
  var path = _ref.path,
      tree = _ref.tree,
      url = _ref.url,
      selected = _ref.selected,
      pathSelected = _ref.pathSelected,
      onBlob = _ref.onBlob,
      depth = _ref.depth,
      filepath = _ref.filepath,
      comparer = _ref.comparer;
  var classes = useStyles();

  var _filepath = _path.default.join(filepath || '', path);

  var icon = selected ? /*#__PURE__*/_react.default.createElement(_icons.Folder, null) : /*#__PURE__*/_react.default.createElement(_icons.FolderOpen, null);
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_core.ListItem, {
    button: true,
    selected: selected,
    className: classes.root,
    style: {
      paddingLeft: depth + 'em'
    }
  }, /*#__PURE__*/_react.default.createElement(_core.ListItemIcon, {
    style: {
      marginRight: 0
    }
  }, icon), /*#__PURE__*/_react.default.createElement(_core.ListItemText, {
    className: classes.pathText,
    primary: path + '/'
  })), /*#__PURE__*/_react.default.createElement(_.Tree, {
    pathSelected: pathSelected,
    tree: tree,
    url: url,
    selected: selected,
    onBlob: onBlob,
    depth: depth + 1,
    filepath: _filepath,
    comparer: comparer
  }));
}

TreeObject.propTypes = {
  /** The filename or path in the Git Tree Object */
  path: _propTypes.default.string.isRequired,

  /** An array of paths from the Gitea file tree api. */
  tree: _propTypes.default.arrayOf(_propTypes.default.shape({
    path: _propTypes.default.string.isRequired,
    type: _propTypes.default.oneOf(['tree', 'blob']).isRequired
  })),

  /** The Url to fetch the listing if listing is not provided. */
  url: _propTypes.default.string,

  /** Set whether or not the File object is currently selected. */
  selected: _propTypes.default.bool,

  /** Set which of the File object is currently selected. */
  pathSelected: _propTypes.default.string,

  /** Function to propogate when the Blob is selected. */
  onBlob: _propTypes.default.func,

  /** The depth of the path in the tree sets the inset of the component. */
  depth: _propTypes.default.number,

  /** The nested filepath that will concatenate. */
  filepath: _propTypes.default.string
};
TreeObject.defaultProps = {
  selected: false,
  depth: 1
};
var _default = TreeObject;
exports.default = _default;
