// gitea-react-toolkit/dist/components/tree-blob/useBlob.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _deepFreeze = _interopRequireDefault(require("deep-freeze"));

var _helpers = require("../file/helpers");

var _ = require(".");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function useBlob(_ref) {
  var _repository$catalog, _repository$catalog$p;

  var config = _ref.config,
      __blob = _ref.blob,
      onBlob = _ref.onBlob,
      repository = _ref.repository,
      tree = _ref.tree,
      _url = _ref.url,
      releaseFlag = _ref.releaseFlag;

  var blob = __blob && (0, _deepFreeze.default)(__blob);

  var url = _url || repository && repository.tree_url;
  console.log("releaseFlag?", releaseFlag);

  if (releaseFlag && repository !== null && repository !== void 0 && (_repository$catalog = repository.catalog) !== null && _repository$catalog !== void 0 && (_repository$catalog$p = _repository$catalog.prod) !== null && _repository$catalog$p !== void 0 && _repository$catalog$p.branch_or_tag_name) {
    // p.replace(regex, 'ferret')
    url = url.replace(/master$/, repository.catalog.prod.branch_or_tag_name);
  }

  console.log("url=", url);
  var update = (0, _react.useCallback)(function (_blob) {
    if (onBlob) {
      onBlob(_blob);
    }

    ;
  }, [onBlob]);
  var close = (0, _react.useCallback)(function () {
    update();
  }, [update]);
  var tsvManifestFileComparer = (0, _react.useCallback)(function (item1, item2) {
    return (0, _helpers.manifestFileComparer)({
      repository: repository,
      item1: item1,
      item2: item2
    });
  }, [repository, repository === null || repository === void 0 ? void 0 : repository.books, _helpers.manifestFileComparer]);
  var browse = (0, _react.useMemo)(function () {
    return tree || url ? /*#__PURE__*/_react.default.createElement(_.Tree, {
      tree: tree,
      url: url,
      config: config,
      selected: true,
      onBlob: update,
      comparer: tsvManifestFileComparer
    }) : /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null);
  }, [tree, url, config, update, repository === null || repository === void 0 ? void 0 : repository.name]);
  return {
    state: blob,
    actions: {
      close: close,
      update: update
    },
    component: browse,
    components: {
      browse: browse
    }
  };
}

;
useBlob.propTypes = {
  /** An array of paths from the Gitea file tree api. */
  tree: _propTypes.default.arrayOf(_propTypes.default.shape({
    path: _propTypes.default.string.isRequired,
    type: _propTypes.default.oneOf(['tree', 'blob']).isRequired
  })),

  /** The Url to fetch the listing if listing is not provided. */
  url: _propTypes.default.string,

  /** Repository tree_url can be used in place of blobConfig */
  repository: _propTypes.default.shape({
    /** The Url to fetch the listing if listing is not provided. */
    tree_url: _propTypes.default.string.isRequired
  })
};
var _default = useBlob;
exports.default = _default;
