// gitea-react-toolkit/dist/components/repository/crud/FormCheckbox.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _core = require("@material-ui/core");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function FormCheckbox(_ref) {
  var name = _ref.name,
      label = _ref.label,
      checked = _ref.checked,
      onChange = _ref.onChange,
      disabled = _ref.disabled;
  return /*#__PURE__*/_react.default.createElement(_core.FormControlLabel, {
    label: label,
    control: /*#__PURE__*/_react.default.createElement(_core.Checkbox, {
      color: "primary",
      value: "private",
      defaultChecked: checked,
      disabled: disabled,
      name: name,
      id: name + Math.random(),
      onChange: onChange
    })
  });
}

FormCheckbox.propTypes = {
  /** The name/id of the field, must be unique. */
  name: _propTypes.default.string.isRequired,

  /** The label of the field. */
  label: _propTypes.default.string.isRequired,

  /** Prepopulate the default value. */
  checked: _propTypes.default.bool,

  /** The function to propogate changes. */
  onChange: _propTypes.default.func,

  /** Set if this field should be required. */
  required: _propTypes.default.bool,

  /** Set if this field should be disabled. */
  disabled: _propTypes.default.bool
};
var _default = FormCheckbox;
exports.default = _default;
