// gitea-react-toolkit/dist/components/repositories/Repositories.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _core = require("@material-ui/core");

var _ = require("../");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function Repositories(_ref) {
  var urls = _ref.urls,
      repositories = _ref.repositories,
      onRepository = _ref.onRepository,
      config = _ref.config;
  var updateRepository = (0, _react.useCallback)(function (repo) {
    onRepository(repo);
  }, [onRepository]);
  var components = [];

  if (repositories) {
    components = repositories.map(function (repository) {
      return /*#__PURE__*/_react.default.createElement(_.Repository, {
        key: JSON.stringify(repository),
        repository: repository,
        onRepository: updateRepository,
        config: config
      });
    });
  } else if (urls) {
    components = urls.map(function (url, index) {
      return /*#__PURE__*/_react.default.createElement(_.Repository, {
        key: index,
        url: url,
        onRepository: updateRepository,
        config: config
      });
    });
  }

  return /*#__PURE__*/_react.default.createElement(_core.List, null, components);
}

Repositories.propTypes = {
  /** Urls array to get repository data, if repository data is not provided. */
  urls: _propTypes.default.array,

  /** Repositories data array to render, if urls not provided. */
  repositories: _propTypes.default.array,

  /** Function to call when repository is selected. */
  onRepository: _propTypes.default.func.isRequired,

  /** Configuration required if paths are provided as URL. */
  config: _propTypes.default.shape({
    server: _propTypes.default.string.isRequired
  })
};
var _default = Repositories;
exports.default = _default;
