// uw-content-validation/dist/core/text-handling-functions.js
"use strict";Object.defineProperty(exports, "__esModule", { value: true });exports.isWhitespace = isWhitespace;exports.countOccurrences = countOccurrences;exports.ourReplaceAll = ourReplaceAll;exports.ourDeleteAll = ourDeleteAll;exports.LEADING_ZERO_COMBINATIONS = exports.BAD_CHARACTER_COMBINATIONS = exports.OPEN_CLOSE_PUNCTUATION_PAIRS = exports.PAIRED_PUNCTUATION_CLOSERS = exports.PAIRED_PUNCTUATION_OPENERS = void 0; // See http://xahlee.info/comp/unicode_matching_brackets.html for a more complete list
// Also see https://unicode-table.com/en/sets/quotation-marks/
var PAIRED_PUNCTUATION_OPENERS = '[({<⟨“‘«‹《〈⸢⸤'; // These are just single/individual characters
exports.PAIRED_PUNCTUATION_OPENERS = PAIRED_PUNCTUATION_OPENERS;var PAIRED_PUNCTUATION_CLOSERS = '])}>⟩”’»›》〉⸣⸥';exports.PAIRED_PUNCTUATION_CLOSERS = PAIRED_PUNCTUATION_CLOSERS;
var OPEN_CLOSE_PUNCTUATION_PAIRS = [
// These can be multiple character sequences (as long as opener and closer are different/distinguishable)
['[', ']'], ['(', ')'], ['{', '}'],
['“', '”'], ['‘', '’'],
['<', '>'], ['⟨', '⟩'],
['«', '»'], ['‹', '›'],
['《', '》'], ['〈', '〉'],
['⸢', '⸣'], ['⸤', '⸥'],
['**_', '_**'] // TODO: Does this markdown belong here???
];exports.OPEN_CLOSE_PUNCTUATION_PAIRS = OPEN_CLOSE_PUNCTUATION_PAIRS;
var BAD_CHARACTER_COMBINATIONS = [
'\\[\\[', '\\]\\]', // These were introduced by a tC Create bug (NOTE: \[ or \] is quite legal)
'] (http', '] (.' // Bad markdown links (with a space between the parts)
];exports.BAD_CHARACTER_COMBINATIONS = BAD_CHARACTER_COMBINATIONS;
var LEADING_ZERO_COMBINATIONS = [
' 0', ':0', '<br>0', '“0', '‘0'];exports.LEADING_ZERO_COMBINATIONS = LEADING_ZERO_COMBINATIONS;


function isWhitespace(myString) {
  // includes zero-width space
  if (/^[\s\u200B]+$/.test(myString)) return true;
  return false;
}


function countOccurrences(mainString, subString) {var allowOverlapping = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : false;
  // Adapted from https://stackoverflow.com/questions/4009756/how-to-count-string-occurrence-in-string

  mainString += '';
  subString += '';
  if (subString.length <= 0) return mainString.length + 1;

  var n = 0,
  pos = 0,
  step = allowOverlapping ? 1 : subString.length;

  while (true) {
    pos = mainString.indexOf(subString, pos);
    if (pos >= 0) {
      ++n;
      pos += step;
    } else break;
  }
  return n;
}
// end of countOccurrences function


function ourReplaceAll(givenString, findString, replaceString) {
  var resultString = givenString;
  while (resultString.indexOf(findString) >= 0) {
    resultString = resultString.replace(findString, replaceString);}
  return resultString;
}


function ourDeleteAll(givenString, findString) {
  var resultString = givenString;
  while (resultString.indexOf(findString) >= 0) {
    resultString = resultString.replace(findString, '');}
  return resultString;
}
