// datatable-translatable/dist/components/toolbar/Toolbar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styles = require("@material-ui/core/styles");

var _core = require("@material-ui/core");

var _icons = require("@material-ui/icons");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function Toolbar(_ref) {
  var preview = _ref.preview,
      onPreview = _ref.onPreview,
      changed = _ref.changed,
      onSave = _ref.onSave,
      onValidate = _ref.onValidate;
  var classes = useStyles();
  var saveDisabled = !changed;
  return /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, onValidate && /*#__PURE__*/_react["default"].createElement(_core.Tooltip, {
    title: "Validate"
  }, /*#__PURE__*/_react["default"].createElement(_core.IconButton, {
    className: classes.iconButton,
    onClick: onValidate,
    "aria-label": "Validate"
  }, /*#__PURE__*/_react["default"].createElement(_icons.PlaylistAddCheck, null))), /*#__PURE__*/_react["default"].createElement(_core.Tooltip, {
    title: "Preview"
  }, /*#__PURE__*/_react["default"].createElement(_core.IconButton, {
    className: classes.iconButton,
    onClick: onPreview,
    "aria-label": "Preview"
  }, preview ? /*#__PURE__*/_react["default"].createElement(_icons.PageviewOutlined, null) : /*#__PURE__*/_react["default"].createElement(_icons.Pageview, null))), /*#__PURE__*/_react["default"].createElement(_core.Tooltip, {
    title: "Save"
  }, /*#__PURE__*/_react["default"].createElement("div", {
    className: classes.inlineDiv
  }, /*#__PURE__*/_react["default"].createElement(_core.IconButton, {
    disabled: saveDisabled,
    className: classes.iconButton,
    onClick: onSave,
    "aria-label": "Save"
  }, changed ? /*#__PURE__*/_react["default"].createElement(_icons.Save, null) : /*#__PURE__*/_react["default"].createElement(_icons.SaveOutlined, null)))));
}

var useStyles = (0, _styles.makeStyles)(function (theme) {
  return {
    iconButton: {},
    inlineDiv: {
      display: 'inline-block'
    }
  };
});
Toolbar.propTypes = {
  /** Current state of preview toggle */
  preview: _propTypes["default"].bool,

  /** Handle click of Preview Button  */
  onPreview: _propTypes["default"].func.isRequired,

  /** Has the file changed for Save to be enabled */
  changed: _propTypes["default"].bool,

  /** Handle click of Save Button */
  onSave: _propTypes["default"].func.isRequired
};
var _default = Toolbar;
exports["default"] = _default;
