// datatable-translatable/dist/components/cell/HeaderCell.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _lodash = _interopRequireDefault(require("lodash.isequal"));

var _ActionsMenu = _interopRequireDefault(require("../actions-menu/ActionsMenu"));

var _styles = _interopRequireDefault(require("./styles"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function HeaderCell(props) {
  var _props$tableMeta = props.tableMeta,
      rowIndex = _props$tableMeta.rowIndex,
      rowData = _props$tableMeta.rowData,
      rowHeader = props.rowHeader,
      delimiters = props.delimiters,
      _props$generateRowId = props.generateRowId,
      generateRowId = _props$generateRowId === void 0 ? function () {} : _props$generateRowId;
  var classes = (0, _styles["default"])();
  var rowHeaderComponent = rowHeader(rowData.slice(1), /*#__PURE__*/_react["default"].createElement(_ActionsMenu["default"], {
    generateRowId: generateRowId,
    rowData: rowData,
    rowIndex: rowIndex,
    delimiters: delimiters
  }));
  return /*#__PURE__*/_react["default"].createElement("div", {
    className: 'header-row ' + classes.root,
    id: generateRowId(rowData)
  }, /*#__PURE__*/_react["default"].createElement("div", {
    className: classes.rowHeader
  }, rowHeaderComponent));
}

;
HeaderCell.propTypes = {
  /** The tableMeta passed from MUIDataTables */
  tableMeta: _propTypes["default"].object.isRequired,

  /** The function to render the rowHeader */
  rowHeader: _propTypes["default"].func,

  /** Set html preview mode, false renders raw markdown */
  generateRowId: _propTypes["default"].func,

  /** Create the row Id for scrolling */
  delimiters: _propTypes["default"].shape({
    /** Delimiters to convert a files into rows "\n" */
    row: _propTypes["default"].string.isRequired,

    /** Delimiters to convert a row into cells "\t" */
    cell: _propTypes["default"].string.isRequired
  }).isRequired
};
HeaderCell.defaultProps = {
  delimiters: {
    row: '\n',
    cell: '\t'
  }
};

var shouldReRender = function shouldReRender(prevProps, nextProps) {
  return (0, _lodash["default"])(prevProps.tableMeta, nextProps.tableMeta);
};

var _default = /*#__PURE__*/(0, _react.memo)(HeaderCell, shouldReRender);

exports["default"] = _default;
