// gitea-react-toolkit/dist/components/tree-blob/BlobObject.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _path = _interopRequireDefault(require("path"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styles = require("@material-ui/core/styles");

var _core = require("@material-ui/core");

var _icons = require("@material-ui/icons");

var _helpers = require("./helpers");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var useStyles = (0, _styles.makeStyles)(function (theme) {
  return {
    root: {
      paddingRight: '0.7em'
    },
    pathText: {
      paddingLeft: '0.7em'
    }
  };
});
/**
 * A Blob Component to render a Git Tree blob object.
 */

function BlobObject(_ref) {
  var selected = _ref.selected,
      blob = _ref.blob,
      _ref$blob = _ref.blob,
      path = _ref$blob.path,
      size = _ref$blob.size,
      onBlob = _ref.onBlob,
      depth = _ref.depth,
      filepath = _ref.filepath;
  var classes = useStyles();

  var _filepath = _path.default.join(filepath || '', path);

  var _blob = _objectSpread(_objectSpread({}, blob), {}, {
    filepath: _filepath
  });

  var icon = selected ? /*#__PURE__*/_react.default.createElement(_icons.Note, null) : /*#__PURE__*/_react.default.createElement(_icons.NoteOutlined, null);
  return /*#__PURE__*/_react.default.createElement(_core.ListItem, {
    button: true,
    selected: selected,
    className: classes.root,
    style: {
      paddingLeft: depth + 'em'
    },
    onClick: function onClick() {
      if (onBlob) {
        onBlob(_blob);
      }
    }
  }, /*#__PURE__*/_react.default.createElement(_core.ListItemIcon, {
    style: {
      marginRight: 0
    }
  }, icon), /*#__PURE__*/_react.default.createElement(_core.ListItemText, {
    className: classes.pathText,
    primary: path,
    secondary: (0, _helpers.humanFileSize)(size)
  }));
}

BlobObject.propTypes = {
  /** Blob data to render, if url not provided. */
  blob: _propTypes.default.shape({
    /** The filepath in the Git Tree Blob Object */
    path: _propTypes.default.string.isRequired,

    /** The url in the Git Tree Blob Object */
    url: _propTypes.default.string,

    /** The content size of the Git Tree Blob Object */
    size: _propTypes.default.number
  }),

  /** Set whether or not the File object is currently selected. */
  selected: _propTypes.default.bool,

  /** Function to propogate when the Blob is selected. */
  onBlob: _propTypes.default.func,

  /** The depth of the path in the tree sets the inset of the component. */
  depth: _propTypes.default.number,

  /** The nested filepath that will concatenate. */
  filepath: _propTypes.default.string
};
BlobObject.defaultProps = {
  selected: false,
  depth: 1
};
var _default = BlobObject;
exports.default = _default;
