// gitea-react-toolkit/dist/components/organization/useOrganization.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _ = require("../..");

var _core = require("../../core");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function useOrganization(_ref) {
  var organization = _ref.organization,
      config = _ref.config,
      onOrganization = _ref.onOrganization,
      authentication = _ref.authentication;
  var update = (0, _react.useCallback)(function (_organization) {
    onOrganization(_organization);
  }, [onOrganization]);
  var list = (0, _react.useCallback)( /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee() {
    var _organizations;

    return regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            _context.next = 2;
            return (0, _core.getCurrentUserOrgs)({
              config: config
            });

          case 2:
            _organizations = _context.sent;
            return _context.abrupt("return", _organizations);

          case 4:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  })), [config]);
  var close = (0, _react.useCallback)(function () {
    update();
  }, [update]);
  var componentProps = {
    config: config || authentication.config,
    authentication: authentication,
    organization: organization,
    onOrganization: update
  };
  var components = {
    view: /*#__PURE__*/_react.default.createElement(_.Organization, componentProps),
    list: /*#__PURE__*/_react.default.createElement(_.CurrentUserOrganizations, componentProps)
  };
  var component = organization ? components.view : components.list;
  return {
    state: organization,
    actions: {
      update: update,
      list: list,
      close: close
    },
    component: component,
    components: components,
    config: config,
    authentication: authentication
  };
}

;
useOrganization.propTypes = {
  /** Organization data to render, if url not provided. */
  organization: _propTypes.default.shape({
    avatar_url: _propTypes.default.string.isRequired,
    description: _propTypes.default.string.isRequired,
    full_name: _propTypes.default.string.isRequired,
    id: _propTypes.default.number.isRequired,
    location: _propTypes.default.string.isRequired,
    username: _propTypes.default.string.isRequired,
    visibility: _propTypes.default.string.isRequired,
    website: _propTypes.default.string.isRequired
  }),

  /** Function to call when organization is selected. */
  onOrganization: _propTypes.default.func.isRequired,

  /** Pass a previously returned authentication object to bypass login. */
  authentication: _propTypes.default.shape({
    user: _propTypes.default.object.isRequired,
    token: _propTypes.default.object.isRequired,
    config: _propTypes.default.object.isRequired,
    remember: _propTypes.default.bool
  }),

  /** Organizations data array to render, if urls not provided. */
  organizations: _propTypes.default.array,

  /** Configuration to pass through to the Search/Organizations component. */
  config: _propTypes.default.shape({
    /** Configuration required for Search or Organizations if paths are provided as URL. */
    server: _propTypes.default.string
  }).isRequired
};
var _default = useOrganization;
exports.default = _default;
