// gitea-react-toolkit/dist/components/file/FileForm.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _core = require("@material-ui/core");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function FileForm(_ref) {
  var _branch = _ref.branch,
      _filepath = _ref.filepath,
      _defaultContent = _ref.defaultContent,
      submitText = _ref.submitText,
      onSubmit = _ref.onSubmit;

  var _useState = (0, _react.useState)(_branch),
      _useState2 = _slicedToArray(_useState, 2),
      branch = _useState2[0],
      setBranch = _useState2[1];

  var _useState3 = (0, _react.useState)(_filepath),
      _useState4 = _slicedToArray(_useState3, 2),
      filepath = _useState4[0],
      setFilepath = _useState4[1];

  var _useState5 = (0, _react.useState)(_defaultContent),
      _useState6 = _slicedToArray(_useState5, 2),
      defaultContent = _useState6[0],
      setDefaultContent = _useState6[1];

  var disabled = !filepath;
  return /*#__PURE__*/_react.default.createElement(_core.Paper, {
    style: {
      marginBottom: '1em',
      padding: '1.3em'
    }
  }, /*#__PURE__*/_react.default.createElement("form", null, /*#__PURE__*/_react.default.createElement("button", {
    type: "submit",
    disabled: true,
    style: {
      display: 'none'
    },
    "aria-hidden": "true"
  }), /*#__PURE__*/_react.default.createElement(_core.TextField, {
    name: "branch",
    label: "branch",
    type: "text",
    autoComplete: null,
    variant: "outlined",
    margin: "normal",
    fullWidth: true,
    defaultValue: branch,
    onChange: function onChange(e) {
      return setBranch(e.target.value);
    }
  }), /*#__PURE__*/_react.default.createElement(_core.TextField, {
    name: "filepath",
    label: "filepath",
    type: "text",
    autoComplete: null,
    variant: "outlined",
    margin: "normal",
    fullWidth: true,
    defaultValue: filepath,
    onChange: function onChange(e) {
      return setFilepath(e.target.value);
    }
  }), /*#__PURE__*/_react.default.createElement(_core.TextField, {
    name: "defaultContent",
    label: "defaultContent",
    type: "text",
    multiline: true,
    autoComplete: null,
    variant: "outlined",
    margin: "normal",
    fullWidth: true,
    defaultValue: defaultContent,
    onChange: function onChange(e) {
      return setDefaultContent(e.target.value);
    }
  }), /*#__PURE__*/_react.default.createElement(_core.Button, {
    type: "button",
    disabled: disabled,
    fullWidth: true,
    variant: "contained",
    color: "primary",
    onClick: function onClick() {
      return onSubmit({
        branch: branch,
        filepath: filepath,
        defaultContent: defaultContent
      });
    }
  }, submitText)));
}

;
FileForm.propTypes = {
  /** text to display on the submit button */
  submitText: _propTypes.default.string,

  /** Function run when submit button is clicked */
  onSubmit: _propTypes.default.func.isRequired
};
FileForm.defaultProps = {
  submitText: 'Submit'
};
var _default = FileForm;
exports.default = _default;
