// gitea-react-toolkit/dist/components/file/FileCard.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styles = require("@material-ui/core/styles");

var _core = require("@material-ui/core");

var _icons = require("@material-ui/icons");

var _markdownTranslatable = require("markdown-translatable");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var useStyles = (0, _styles.makeStyles)(function (theme) {
  return {
    actions: {
      textAlign: 'right',
      paddingRight: "".concat(theme.spacing(1.5), "px"),
      width: '100%'
    },
    action: {
      marginLeft: "0px"
    }
  };
});

function FileCard(_ref) {
  var authentication = _ref.authentication,
      repository = _ref.repository,
      file = _ref.file;
  var classes = useStyles();

  var _useState = (0, _react.useState)(true),
      _useState2 = _slicedToArray(_useState, 2),
      preview = _useState2[0],
      setPreview = _useState2[1];

  var _useState3 = (0, _react.useState)(file ? file.content : ''),
      _useState4 = _slicedToArray(_useState3, 2),
      markdown = _useState4[0],
      setMarkdown = _useState4[1];

  var changed = markdown !== (file && file.content);
  var avatarUrl = repository.avatar_url || repository.owner.avatar_url;
  var access = repository.permissions.push;
  (0, _react.useEffect)(function () {
    setMarkdown(file && file.content);
  }, [file]);
  var branch = file && file.branch ? file.branch : repository.default_branch;
  return /*#__PURE__*/_react.default.createElement(_core.Card, null, /*#__PURE__*/_react.default.createElement(_core.CardHeader, {
    avatar: /*#__PURE__*/_react.default.createElement(_core.Avatar, {
      src: avatarUrl
    }),
    title: /*#__PURE__*/_react.default.createElement("strong", null, file && file.path),
    subheader: repository.full_name + '/' + branch
  }), /*#__PURE__*/_react.default.createElement(_core.CardContent, null, /*#__PURE__*/_react.default.createElement(_core.Paper, null, /*#__PURE__*/_react.default.createElement(_markdownTranslatable.BlockEditable, {
    preview: preview,
    markdown: markdown,
    onEdit: setMarkdown,
    editable: !!authentication
  }))), /*#__PURE__*/_react.default.createElement(_core.CardActions, null, /*#__PURE__*/_react.default.createElement("div", {
    className: classes.actions
  }, /*#__PURE__*/_react.default.createElement(_core.IconButton, {
    className: classes.action,
    "aria-label": "Preview",
    onClick: function onClick() {
      return setPreview(!preview);
    }
  }, !preview ? /*#__PURE__*/_react.default.createElement(_icons.Pageview, null) : /*#__PURE__*/_react.default.createElement(_icons.PageviewOutlined, null)), /*#__PURE__*/_react.default.createElement(_core.IconButton, {
    className: classes.action,
    "aria-label": "Save",
    disabled: !access,
    onClick: function onClick() {
      if (changed) file.save(markdown);
    }
  }, changed ? /*#__PURE__*/_react.default.createElement(_icons.Save, null) : /*#__PURE__*/_react.default.createElement(_icons.SaveOutlined, null)), /*#__PURE__*/_react.default.createElement(_core.IconButton, {
    className: classes.action,
    "aria-label": "Delete",
    disabled: !access,
    onClick: function onClick() {
      var confirmation = window.confirm("Are you sure you want to Delete ".concat(file.filepath, "?"));
      if (confirmation) file.dangerouslyDelete();
    }
  }, /*#__PURE__*/_react.default.createElement(_icons.DeleteSweepOutlined, null)), /*#__PURE__*/_react.default.createElement(_core.IconButton, {
    title: "Close Repository",
    onClick: file.close
  }, /*#__PURE__*/_react.default.createElement(_icons.CancelOutlined, null)))));
}

;
FileCard.propTypes = {
  /** Repository tree_url can be used in place of blobConfig */
  repository: _propTypes.default.shape({
    owner: _propTypes.default.shape({
      username: _propTypes.default.string.isRequired,
      avatar_url: _propTypes.default.string.isRequired
    }),
    name: _propTypes.default.string.isRequired,
    avatar_url: _propTypes.default.string
  }).isRequired,

  /** Pass a previously returned file object to bypass the selection. */
  file: _propTypes.default.shape({
    name: _propTypes.default.string.isRequired,
    path: _propTypes.default.string.isRequired,
    sha: _propTypes.default.string.isRequired,
    content: _propTypes.default.string,
    branch: _propTypes.default.string,
    filepath: _propTypes.default.string
  }),

  /** Pass a previously returned authentication object to bypass login. */
  authentication: _propTypes.default.shape({
    user: _propTypes.default.object.isRequired,
    token: _propTypes.default.object.isRequired,
    config: _propTypes.default.object.isRequired
  })
};
var _default = FileCard;
exports.default = _default;
