// gitea-react-toolkit/dist/components/file/useEdit.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = useEdit;

var _react = require("react");

var _ = require("../..");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function useEdit(_ref) {
  var sha = _ref.sha,
      repo = _ref.repo,
      owner = _ref.owner,
      token = _ref.token,
      config = _ref.config,
      branch = _ref.branch,
      author = _ref.author,
      content = _ref.content,
      message = _ref.message,
      filepath = _ref.filepath;

  var _useState = (0, _react.useState)({
    editResponse: null,
    isEditing: false,
    isError: false,
    error: null
  }),
      _useState2 = _slicedToArray(_useState, 2),
      _useState2$ = _useState2[0],
      isEditing = _useState2$.isEditing,
      isError = _useState2$.isError,
      error = _useState2$.error,
      editResponse = _useState2$.editResponse,
      setState = _useState2[1];

  var _ref2 = token || {},
      tokenid = _ref2.name;

  var _message = message || "Edit '".concat(filepath, "' using '").concat(tokenid, "'");

  function onSaveEdit(_x) {
    return _onSaveEdit.apply(this, arguments);
  }

  function _onSaveEdit() {
    _onSaveEdit = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee(_branch) {
      var response;
      return regeneratorRuntime.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              _context.prev = 0;

              if (!content) {
                _context.next = 10;
                break;
              }

              setState(function (prevState) {
                return _objectSpread(_objectSpread({}, prevState), {}, {
                  editResponse: null,
                  isEditing: true,
                  isError: false
                });
              });
              _context.next = 5;
              return (0, _.updateContent)({
                sha: sha,
                repo: repo,
                owner: owner,
                config: config,
                author: author,
                content: content,
                filepath: filepath,
                message: _message,
                // Use branch passed to function or branch passed to custom hook. 
                branch: _branch || branch
              });

            case 5:
              response = _context.sent;
              setState(function (prevState) {
                return _objectSpread(_objectSpread({}, prevState), {}, {
                  editResponse: response
                });
              });
              return _context.abrupt("return", true);

            case 10:
              console.warn('Content value is empty');

            case 11:
              _context.next = 17;
              break;

            case 13:
              _context.prev = 13;
              _context.t0 = _context["catch"](0);
              setState(function (prevState) {
                return _objectSpread(_objectSpread({}, prevState), {}, {
                  isError: true,
                  error: _context.t0
                });
              });
              return _context.abrupt("return", false);

            case 17:
              _context.prev = 17;
              setState(function (prevState) {
                return _objectSpread(_objectSpread({}, prevState), {}, {
                  isEditing: false
                });
              });
              return _context.finish(17);

            case 20:
            case "end":
              return _context.stop();
          }
        }
      }, _callee, null, [[0, 13, 17, 20]]);
    }));
    return _onSaveEdit.apply(this, arguments);
  }

  return {
    error: error,
    isError: isError,
    isEditing: isEditing,
    onSaveEdit: onSaveEdit,
    editResponse: editResponse
  };
}
