// gitea-react-toolkit/dist/components/file/useFile.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _useDeepCompareEffect = _interopRequireDefault(require("use-deep-compare-effect"));

var _helpers = require("./helpers");

var _ = require("..");

var _dcsCatalogNextApis = require("./dcsCatalogNextApis");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function useFile(_ref) {
  var authentication = _ref.authentication,
      repository = _ref.repository,
      filepath = _ref.filepath,
      onFilepath = _ref.onFilepath,
      defaultContent = _ref.defaultContent,
      _config = _ref.config,
      _ref$create = _ref.create,
      create = _ref$create === void 0 ? false : _ref$create,
      onOpenValidation = _ref.onOpenValidation,
      onLoadCache = _ref.onLoadCache,
      onSaveCache = _ref.onSaveCache,
      onConfirmClose = _ref.onConfirmClose,
      releaseFlag = _ref.releaseFlag;

  var _useState = (0, _react.useState)(),
      _useState2 = _slicedToArray(_useState, 2),
      file = _useState2[0],
      setFile = _useState2[1];

  var _useState3 = (0, _react.useState)(false),
      _useState4 = _slicedToArray(_useState3, 2),
      isChanged = _useState4[0],
      setIsChanged = _useState4[1];

  var _useState5 = (0, _react.useState)(),
      _useState6 = _slicedToArray(_useState5, 2),
      blob = _useState6[0],
      setBlob = _useState6[1];

  var _useContext = (0, _react.useContext)(_.RepositoryContext),
      updateBranch = _useContext.actions.updateBranch,
      repositoryConfig = _useContext.config;

  var config = _config || repositoryConfig;
  var branch = repository && (repository.branch || repository.default_branch);

  var _useState7 = (0, _react.useState)(),
      _useState8 = _slicedToArray(_useState7, 2),
      deleted = _useState8[0],
      setDeleted = _useState8[1];

  var _setBlob = (0, _react.useCallback)(function (_blob) {
    if (blob && _blob && onConfirmClose) {
      if (onConfirmClose()) {
        setBlob(_blob);
      }
    } else {
      setBlob(_blob);
    }
  }, [blob, setBlob, onConfirmClose]);

  var _useBlob = (0, _.useBlob)({
    blob: blob,
    onBlob: _setBlob,
    config: config,
    repository: repository,
    filepath: filepath,
    releaseFlag: releaseFlag
  }),
      blobState = _useBlob.state,
      blobActions = _useBlob.actions,
      blobComponents = _useBlob.components;

  var _ref2 = repository && repository.permissions ? repository.permissions : {},
      writeable = _ref2.push;

  var update = (0, _react.useCallback)(function (_file) {
    setFile(_file);
  }, []);
  (0, _react.useEffect)(function () {
    setIsChanged(false);
  }, [file, deleted, closed]);
  var read = (0, _react.useCallback)( /*#__PURE__*/function () {
    var _ref3 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee(_filepath) {
      return regeneratorRuntime.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              if (!onFilepath) {
                _context.next = 3;
                break;
              }

              _context.next = 3;
              return onFilepath(_filepath);

            case 3:
              ;

            case 4:
            case "end":
              return _context.stop();
          }
        }
      }, _callee);
    }));

    return function (_x) {
      return _ref3.apply(this, arguments);
    };
  }(), [onFilepath]);
  var load = (0, _react.useCallback)( /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee2() {
    var _file, defaultCachedContentFile, content, _publishedContent, _repository$catalog, _repository$catalog$p, prodTag;

    return regeneratorRuntime.wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            if (!(config && repository && filepath)) {
              _context2.next = 29;
              break;
            }

            _context2.next = 3;
            return (0, _helpers.ensureFile)({
              filepath: filepath,
              defaultContent: defaultContent,
              authentication: authentication,
              config: config,
              repository: repository,
              branch: branch,
              onOpenValidation: onOpenValidation
            });

          case 3:
            _file = _context2.sent;
            console.log("ensureFile:");
            console.log(_file);
            // if (loadDefaultCachedContentFile) { 
            //   console.log("_file.html_url");
            //   console.log(_file.html_url);
            //   defaultCachedContentFile = await loadDefaultCachedContentFile(_file.html_url);
            // }
            console.log("GRT // onLoadCache!", _file, _file.html_url);

            if (!(onLoadCache && _file && _file.html_url)) {
              _context2.next = 11;
              break;
            }

            _context2.next = 10;
            return onLoadCache({
              authentication: authentication,
              repository: repository,
              branch: branch,
              html_url: _file.html_url
            });

          case 10:
            defaultCachedContentFile = _context2.sent;

          case 11:
            console.log("GRT defaultContent", '|', defaultContent);
            console.log("GRT defaultCachedContent", '|', defaultCachedContentFile);

            if (defaultCachedContentFile) {
              console.log(defaultCachedContentFile.sha);
            }

            if (_file) {
              console.log(_file.sha);
            } // let content;
            // content = await repositoryActions.fileFromZip(filepath);
            //const content = await getContentFromFile(_file);
            // Allow app to provide CACHED ("offline" content);
            // Might be different BRANCH (different user) or different FILE.
            // Might be STALE (sha has changed on DCS).
            // (NOTE: STALE cache would mean THIS user edited the same file in another browser.)


            if (!(defaultCachedContentFile && defaultCachedContentFile.content && defaultCachedContentFile.sha && defaultCachedContentFile.html_url && defaultCachedContentFile.filepath && defaultCachedContentFile.timestamp && defaultCachedContentFile.sha === _file.sha && defaultCachedContentFile.html_url === _file.html_url)) {
              _context2.next = 19;
              break;
            }

            // Load autosaved content:
            content = defaultCachedContentFile.content;
            _context2.next = 28;
            break;

          case 19:
            if (_file && _file.content && defaultCachedContentFile && defaultCachedContentFile.content && defaultCachedContentFile.filepath && defaultCachedContentFile.sha && defaultCachedContentFile.timestamp) {
              console.log(_file);
              console.log(defaultCachedContentFile);
              alert("A previous file was autosaved. The autosaved file will be overwritten.\n\n" + "File: " + defaultCachedContentFile.filepath + ".\n" + "Edited: " + defaultCachedContentFile.timestamp.toLocaleString() + ".");
            } // Get SERVER content: Overwrite cache:


            _context2.next = 22;
            return (0, _helpers.getContentFromFile)(_file);

          case 22:
            content = _context2.sent;
            // Check catalog next:
            prodTag = (_repository$catalog = repository.catalog) === null || _repository$catalog === void 0 ? void 0 : (_repository$catalog$p = _repository$catalog.prod) === null || _repository$catalog$p === void 0 ? void 0 : _repository$catalog$p.branch_or_tag_name;

            if (!prodTag) {
              _context2.next = 28;
              break;
            }

            _context2.next = 27;
            return (0, _dcsCatalogNextApis.fetchCatalogContent)('unfoldingword', repository.name, prodTag, filepath, config);

          case 27:
            _publishedContent = _context2.sent;

          case 28:
            update(_objectSpread(_objectSpread({}, _file), {}, {
              branch: branch,
              content: content,
              filepath: _file.path,
              publishedContent: _publishedContent
            }));

          case 29:
          case "end":
            return _context2.stop();
        }
      }
    }, _callee2);
  })), [authentication, branch, config, defaultContent, onLoadCache, filepath, repository, update]);
  var createFile = (0, _react.useCallback)( /*#__PURE__*/function () {
    var _ref6 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee3(_ref5) {
      var _branch, _filepath, _defaultContent, onOpenValidation, _file;

      return regeneratorRuntime.wrap(function _callee3$(_context3) {
        while (1) {
          switch (_context3.prev = _context3.next) {
            case 0:
              _branch = _ref5.branch, _filepath = _ref5.filepath, _defaultContent = _ref5.defaultContent, onOpenValidation = _ref5.onOpenValidation;

              if (!(config && repository)) {
                _context3.next = 7;
                break;
              }

              _context3.next = 4;
              return (0, _helpers.ensureFile)({
                authentication: authentication,
                config: config,
                repository: repository,
                branch: _branch,
                filepath: _filepath,
                defaultContent: _defaultContent,
                onOpenValidation: onOpenValidation
              });

            case 4:
              _file = _context3.sent;

              if (_file) {
                updateBranch(_branch);
                onFilepath(_filepath);
              }

              ;

            case 7:
            case "end":
              return _context3.stop();
          }
        }
      }, _callee3);
    }));

    return function (_x2) {
      return _ref6.apply(this, arguments);
    };
  }(), [authentication, config, repository, updateBranch, onFilepath]);
  var close = (0, _react.useCallback)(function () {
    if (blobActions && blobActions.close) {
      blobActions.close();
    }

    ;

    if (onFilepath) {
      onFilepath();
    }

    ;
    update();
  }, [update, blobActions, onFilepath]);
  var save = (0, _react.useCallback)( /*#__PURE__*/function () {
    var _ref7 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee5(content) {
      return regeneratorRuntime.wrap(function _callee5$(_context5) {
        while (1) {
          switch (_context5.prev = _context5.next) {
            case 0:
              console.log("GRT save // will save file");
              _context5.next = 3;
              return (0, _helpers.saveFile)({
                authentication: authentication,
                repository: repository,
                branch: branch,
                file: file,
                content: content
              }).then(
              /*#__PURE__*/
              // Empty cache if user has saved this file
              // (save() will not happen for "OFFLINE" system files)
              _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee4() {
                return regeneratorRuntime.wrap(function _callee4$(_context4) {
                  while (1) {
                    switch (_context4.prev = _context4.next) {
                      case 0:
                        console.log("GRT save // will EMPTY cache");
                        _context4.next = 3;
                        return saveCache(null);

                      case 3:
                        console.log("GRT save // will load file");
                        _context4.next = 6;
                        return load();

                      case 6:
                      case "end":
                        return _context4.stop();
                    }
                  }
                }, _callee4);
              })));

            case 3:
            case "end":
              return _context5.stop();
          }
        }
      }, _callee5);
    }));

    return function (_x3) {
      return _ref7.apply(this, arguments);
    };
  }(), [writeable, authentication, repository, branch, file, load, _helpers.saveFile, saveCache]);
  var saveCache = (0, _react.useCallback)( /*#__PURE__*/function () {
    var _ref9 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee6(content) {
      return regeneratorRuntime.wrap(function _callee6$(_context6) {
        while (1) {
          switch (_context6.prev = _context6.next) {
            case 0:
              if (!onSaveCache) {
                _context6.next = 3;
                break;
              }

              _context6.next = 3;
              return onSaveCache({
                authentication: authentication,
                repository: repository,
                branch: branch,
                file: file,
                content: content
              });

            case 3:
            case "end":
              return _context6.stop();
          }
        }
      }, _callee6);
    }));

    return function (_x4) {
      return _ref9.apply(this, arguments);
    };
  }(), [writeable, authentication, repository, branch, file, onSaveCache]);
  var dangerouslyDelete = (0, _react.useCallback)( /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee7() {
    var _deleted;

    return regeneratorRuntime.wrap(function _callee7$(_context7) {
      while (1) {
        switch (_context7.prev = _context7.next) {
          case 0:
            if (!writeable) {
              _context7.next = 5;
              break;
            }

            _context7.next = 3;
            return (0, _helpers.deleteFile)({
              authentication: authentication,
              repository: repository,
              file: file,
              branch: branch
            });

          case 3:
            _deleted = _context7.sent;

            if (_deleted) {
              setDeleted(true);
            }

          case 5:
            ;

          case 6:
          case "end":
            return _context7.stop();
        }
      }
    }, _callee7);
  })), [file, authentication, branch, repository, writeable]);
  (0, _useDeepCompareEffect.default)(function () {
    var notLoaded = !file && filepath && !deleted;
    var loadNew = file && filepath && file.filepath !== filepath;

    if (notLoaded || loadNew) {
      load();
    }
  }, [deleted, filepath, load, file]);
  var blobFilepath = blobState && blobState.filepath;
  (0, _react.useEffect)(function () {
    if (blobFilepath && onFilepath) {
      onFilepath(blobFilepath);
    }

    ;
  }, [blobFilepath, onFilepath]);
  (0, _useDeepCompareEffect.default)(function () {
    // if there is a file but no repository, close file.
    if (!repository && file) close();
  }, [repository, file, close]);
  (0, _react.useEffect)(function () {
    if (deleted) {
      close();
      setDeleted(false);
    }

    ;
  }, [deleted, close]);
  var actions = {
    update: update,
    load: load,
    read: read,
    save: save,
    saveCache: saveCache,
    onSaveCache: onSaveCache,
    onLoadCache: onLoadCache,
    close: close,
    dangerouslyDelete: dangerouslyDelete,
    setIsChanged: setIsChanged,
    onConfirmClose: onConfirmClose
  };
  var components = {
    create: repository && /*#__PURE__*/_react.default.createElement(_.FileForm, {
      branch: branch,
      defaultContent: defaultContent,
      onSubmit: createFile
    }),
    browse: repository && blobComponents.browse,
    fileCard: repository && file && /*#__PURE__*/_react.default.createElement(_.FileCard, {
      authentication: authentication,
      repository: repository,
      file: _objectSpread(_objectSpread({}, file), actions)
    })
  };

  var component = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null);

  if (file) {
    component = components.fileCard;
  } else if (!filepath) {
    if (create) {
      component = components.create;
    } else {
      component = components.browse;
    }
  }

  return {
    state: file,
    stateValues: {
      isChanged: isChanged
    },
    actions: actions,
    component: component,
    components: components
  };
}

;
useFile.propTypes = {
  /** The full filepath for the file. */
  filepath: _propTypes.default.string,

  /** Function to propogate filepath when the Blob is selected. */
  onFilepath: _propTypes.default.func,

  /** Authentication object returned from a successful withAuthentication login. */
  authentication: _propTypes.default.shape({
    config: _propTypes.default.shape({
      server: _propTypes.default.string.isRequired,
      headers: _propTypes.default.shape({
        Authorization: _propTypes.default.string.isRequired
      }).isRequired
    }).isRequired,
    user: _propTypes.default.shape({
      username: _propTypes.default.string.isRequired,
      email: _propTypes.default.string.isRequired
    }).isRequired
  }),

  /** Repository tree_url can be used in place of blobConfig */
  repository: _propTypes.default.shape({
    owner: _propTypes.default.shape({
      username: _propTypes.default.string.isRequired
    }),
    name: _propTypes.default.string.isRequired
  }).isRequired,

  /** use a form to create a new file */
  create: _propTypes.default.bool
};
var _default = useFile;
exports.default = _default;
