// gitea-react-toolkit/dist/components/repository/useRepository.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _deepFreeze = _interopRequireDefault(require("deep-freeze"));

var _ = require("../..");

var _helpers = require("./helpers");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function useRepository(_ref) {
  var repositories = _ref.repositories,
      __full_name = _ref.full_name,
      urls = _ref.urls,
      defaultOwner = _ref.defaultOwner,
      defaultQuery = _ref.defaultQuery,
      config = _ref.config,
      authentication = _ref.authentication,
      __repository = _ref.repository,
      onRepository = _ref.onRepository,
      __branch = _ref.branch;

  var _useState = (0, _react.useState)(__branch),
      _useState2 = _slicedToArray(_useState, 2),
      branch = _useState2[0],
      setBranch = _useState2[1];

  var repository = __repository && (0, _deepFreeze.default)(__repository);

  var _ref2 = repository || {},
      full_name = _ref2.full_name;

  var user = authentication && authentication.user; // Due to objects not memoizing in useCallback, deconstruction is necessary
  // const { owner: { username: owner }, name: repo } = repository;

  var update = (0, _react.useCallback)(function (repo) {
    if (onRepository) {
      var _repo;

      if (repo) {
        _repo = _objectSpread(_objectSpread({}, repo), {}, {
          branch: branch
        });
        var tree_url = (0, _.repoTreeUrl)(_repo);
        _repo = _objectSpread(_objectSpread({}, _repo), {}, {
          tree_url: tree_url
        });
      }

      ;
      onRepository(_repo);
    }

    ;
  }, [branch, onRepository]);
  var read = (0, _react.useCallback)( /*#__PURE__*/function () {
    var _ref4 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee(_ref3) {
      var owner, name, _repository;

      return regeneratorRuntime.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              owner = _ref3.owner, name = _ref3.name;
              _context.next = 3;
              return (0, _helpers.readRepository)({
                owner: owner,
                repo: name,
                config: config
              });

            case 3:
              _repository = _context.sent;
              update(_repository);

            case 5:
            case "end":
              return _context.stop();
          }
        }
      }, _callee);
    }));

    return function (_x) {
      return _ref4.apply(this, arguments);
    };
  }(), [config, update]);
  var updateBranch = (0, _react.useCallback)(function (_branch) {
    setBranch(_branch);
  }, []);
  var create = (0, _react.useCallback)( /*#__PURE__*/function () {
    var _ref5 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee2(settings) {
      var _repository;

      return regeneratorRuntime.wrap(function _callee2$(_context2) {
        while (1) {
          switch (_context2.prev = _context2.next) {
            case 0:
              _context2.next = 2;
              return (0, _helpers.createRepository)({
                settings: settings,
                config: config
              });

            case 2:
              _repository = _context2.sent;
              update(_repository);

            case 4:
            case "end":
              return _context2.stop();
          }
        }
      }, _callee2);
    }));

    return function (_x2) {
      return _ref5.apply(this, arguments);
    };
  }(), [config, update]);
  var dangerouslyDelete = (0, _react.useCallback)(function () {
    if (user && user.username === repository.owner.username) {
      (0, _helpers.deleteRepository)({
        repository: repository,
        config: config
      });
      window.setTimeout(update, 500);
    }

    ;
  }, [user, repository, config, update]);
  var fork = (0, _react.useCallback)( /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee3() {
    var _repo;

    return regeneratorRuntime.wrap(function _callee3$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            if (!((user && user.username) !== repository.owner.username)) {
              _context3.next = 5;
              break;
            }

            _context3.next = 3;
            return (0, _helpers.forkRepository)({
              repository: repository,
              config: config
            });

          case 3:
            _repo = _context3.sent;
            update(_repo);

          case 5:
            ;

          case 6:
          case "end":
            return _context3.stop();
        }
      }
    }, _callee3);
  })), [config, repository, update, user]);
  var save = (0, _react.useCallback)( /*#__PURE__*/function () {
    var _ref7 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee4(settings) {
      var _repository;

      return regeneratorRuntime.wrap(function _callee4$(_context4) {
        while (1) {
          switch (_context4.prev = _context4.next) {
            case 0:
              if (!repository.permissions.admin) {
                _context4.next = 5;
                break;
              }

              _context4.next = 3;
              return (0, _helpers.saveRepository)({
                repository: repository,
                settings: settings,
                config: config
              });

            case 3:
              _repository = _context4.sent;
              update(_repository);

            case 5:
              ;

            case 6:
            case "end":
              return _context4.stop();
          }
        }
      }, _callee4);
    }));

    return function (_x3) {
      return _ref7.apply(this, arguments);
    };
  }(), [config, repository, update]);
  var storeZip = (0, _react.useCallback)( /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee5() {
    return regeneratorRuntime.wrap(function _callee5$(_context5) {
      while (1) {
        switch (_context5.prev = _context5.next) {
          case 0:
            _context5.next = 2;
            return (0, _helpers.storeRepositoryZip)({
              repository: repository,
              config: config
            });

          case 2:
          case "end":
            return _context5.stop();
        }
      }
    }, _callee5);
  })), [repository, config]);
  var removeZip = (0, _react.useCallback)( /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee6() {
    return regeneratorRuntime.wrap(function _callee6$(_context6) {
      while (1) {
        switch (_context6.prev = _context6.next) {
          case 0:
            _context6.next = 2;
            return (0, _helpers.removeRepositoryZip)({
              repository: repository,
              config: config
            });

          case 2:
          case "end":
            return _context6.stop();
        }
      }
    }, _callee6);
  })), [repository, config]);
  var fileFromZip = (0, _react.useCallback)( /*#__PURE__*/function () {
    var _ref10 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee7(filepath) {
      var file;
      return regeneratorRuntime.wrap(function _callee7$(_context7) {
        while (1) {
          switch (_context7.prev = _context7.next) {
            case 0:
              _context7.next = 2;
              return (0, _helpers.getFileFromRepositoryZip)({
                repository: repository,
                filepath: filepath,
                config: config
              });

            case 2:
              file = _context7.sent;
              return _context7.abrupt("return", file);

            case 4:
            case "end":
              return _context7.stop();
          }
        }
      }, _callee7);
    }));

    return function (_x4) {
      return _ref10.apply(this, arguments);
    };
  }(), [repository, config]);
  var forks = (0, _react.useCallback)(function () {
    (0, _helpers.repositoryForks)({
      repository: repository,
      config: config
    });
  }, [repository, config]);
  var close = (0, _react.useCallback)(function () {
    update();
  }, [update]);
  (0, _react.useEffect)(function () {
    if (__branch !== branch) {
      setBranch(__branch);
    }
  }, [__branch, branch]);
  (0, _react.useEffect)(function () {
    if (repository && branch !== repository.branch) {
      update(_objectSpread(_objectSpread({}, repository), {}, {
        branch: branch
      }));
    }
  }, [branch, repository, update]);
  (0, _react.useEffect)(function () {
    if (__full_name && !full_name) {
      var _full_name$split = __full_name.split('/'),
          _full_name$split2 = _slicedToArray(_full_name$split, 2),
          owner = _full_name$split2[0],
          name = _full_name$split2[1];

      read({
        owner: owner,
        name: name
      });
    }

    ;
  }, [full_name, __full_name, read]);
  var hasRepository = repository && repository.name && repository.owner && repository.permissions;
  var username = authentication && authentication.user && authentication.user.username;
  var components = {
    search: config && /*#__PURE__*/_react.default.createElement(_.Search, {
      defaultOwner: defaultOwner || username,
      defaultQuery: defaultQuery,
      onRepository: update,
      config: config
    }),
    browse: (urls || repositories) && /*#__PURE__*/_react.default.createElement(_.Repositories, {
      urls: urls,
      repositories: repositories,
      onRepository: update,
      config: config
    }),
    view: hasRepository && /*#__PURE__*/_react.default.createElement(_.Repository, {
      repository: repository,
      config: config,
      onRepository: onRepository
    }),
    form: hasRepository && /*#__PURE__*/_react.default.createElement(_.RepositoryForm, null)
  };

  var component = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null);

  if (hasRepository) {
    component = components.form;
  } else if (urls || repositories) {
    component = components.browse;
  } else if (config) {
    component = components.search;
  }

  ;
  return {
    state: repository,
    actions: {
      close: close,
      create: create,
      update: update,
      dangerouslyDelete: dangerouslyDelete,
      fork: fork,
      save: save,
      forks: forks,
      updateBranch: updateBranch,
      read: read,
      storeZip: storeZip,
      removeZip: removeZip,
      fileFromZip: fileFromZip
    },
    component: component,
    components: components,
    config: config
  };
}

;
useRepository.propTypes = {
  /** Repository data to render, if url not provided. */
  repository: _propTypes.default.shape({
    id: _propTypes.default.number,
    owner: _propTypes.default.object.isRequired,
    name: _propTypes.default.string.isRequired,
    full_name: _propTypes.default.string.isRequired,
    description: _propTypes.default.string.isRequired,
    html_url: _propTypes.default.string.isRequired,
    website: _propTypes.default.string.isRequired,
    tree_url: _propTypes.default.string,
    avatar_url: _propTypes.default.string,
    branch: _propTypes.default.string
  }),

  /** Function to call when repository is selected. */
  onRepository: _propTypes.default.func.isRequired,

  /** Full name of the repo, `owner/name` */
  full_name: _propTypes.default.string,

  /** Pass a previously returned authentication object to bypass login. */
  authentication: _propTypes.default.shape({
    user: _propTypes.default.object.isRequired,
    token: _propTypes.default.object.isRequired,
    config: _propTypes.default.object.isRequired,
    remember: _propTypes.default.bool
  }),

  /** Urls array to get repository data, if repository data is not provided. */
  urls: _propTypes.default.array,

  /** Repositories data array to render, if urls not provided. */
  repositories: _propTypes.default.array,

  /** Prefill the owner search field. */
  defaultOwner: _propTypes.default.string,

  /** Prefill the query search field. */
  defaultQuery: _propTypes.default.string,

  /** The name of the branch to read/write files */
  branch: _propTypes.default.string,

  /** Configuration to pass through to the Search/Repositories component. */
  config: _propTypes.default.shape({
    /** Configuration required for Search or Repositories if paths are provided as URL. */
    server: _propTypes.default.string
  }).isRequired
};
var _default = useRepository;
exports.default = _default;
