// datatable-translatable/dist/components/column-filter/ColumnFilter.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _core = require("@material-ui/core");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function ColumnFilter(_ref) {
  var filterList = _ref.filterList,
      onChange = _ref.onChange,
      filterIndex = _ref.filterIndex,
      column = _ref.column,
      optionValues = _ref.optionValues;

  var handleChange = function handleChange(event) {
    var value = event.target.value;
    if (value) onChange(value, filterIndex, column.name);
  };

  return /*#__PURE__*/_react["default"].createElement(_core.FormControl, {
    key: filterIndex,
    fullWidth: true
  }, /*#__PURE__*/_react["default"].createElement(_core.InputLabel, {
    htmlFor: column.name
  }, column.label), /*#__PURE__*/_react["default"].createElement(_core.Select, {
    fullWidth: true,
    value: filterList[filterIndex].length ? filterList[filterIndex].toString() : 'All',
    name: column.name,
    onChange: handleChange,
    input: /*#__PURE__*/_react["default"].createElement(_core.Input, {
      name: column.name,
      id: column.name
    })
  }, /*#__PURE__*/_react["default"].createElement(_core.MenuItem, {
    value: 'All',
    key: 0
  }, 'All'), optionValues.map(function (filterValue, _filterIndex) {
    return /*#__PURE__*/_react["default"].createElement(_core.MenuItem, {
      value: filterValue,
      key: _filterIndex + 1
    }, filterValue != null ? filterValue.toString() : '');
  })));
}

;
var _default = ColumnFilter;
exports["default"] = _default;
