// gitea-react-toolkit/dist/core/gitea-api/fetch/repositories.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.fetchResourceManifests = fetchResourceManifests;
exports.getLanguages = getLanguages;
exports.getLanguageIds = getLanguageIds;
exports.getLanguageIdsByResource = getLanguageIdsByResource;
exports.fetchManifest = fetchManifest;
exports.fetchRepositoriesZipFiles = fetchRepositoriesZipFiles;
exports.resourceRepositories = void 0;

var _path = _interopRequireDefault(require("path"));

var _jsYamlParser = _interopRequireDefault(require("js-yaml-parser"));

var _ = require("..");

var _languages = require("../../languages");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

var resourceRepositories = function resourceRepositories(_ref) {
  var languageId = _ref.languageId;
  return {
    obs: languageId + '_obs',
    'obs-tn': languageId + '_obs-tn',
    'obs-tq': languageId + '_obs-tq',
    'obs-sn': languageId + '_obs-sn',
    'obs-sq': languageId + '_obs-sq',
    ult: languageId + '_ult',
    ust: languageId + '_ust',
    irv: languageId + '_irv',
    ulb: languageId + '_ulb',
    udb: languageId + '_udb',
    tn: languageId + '_tn',
    ta: languageId + '_ta',
    tw: languageId + '_tw',
    ugnt: 'UGNT',
    uhb: 'UHB',
    ugl: languageId + '_ugl',
    uhal: languageId + '_uhal'
  };
};

exports.resourceRepositories = resourceRepositories;

function fetchResourceManifests(_x) {
  return _fetchResourceManifests.apply(this, arguments);
}

function _fetchResourceManifests() {
  _fetchResourceManifests = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee(_ref2) {
    var username, languageId, manifests, _resourceRepositories, resourceIds, promises, manifestArray;

    return regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            username = _ref2.username, languageId = _ref2.languageId;
            manifests = {};
            _resourceRepositories = resourceRepositories({
              languageId: languageId
            });
            resourceIds = Object.keys(_resourceRepositories);
            promises = resourceIds.map(function (resourceId) {
              var repository = _resourceRepositories[resourceId];

              var _username = ['ugnt', 'uhb'].includes(resourceId) ? 'unfoldingword' : username;

              return fetchManifest({
                username: _username,
                repository: repository
              });
            });
            _context.next = 7;
            return Promise.all(promises);

          case 7:
            manifestArray = _context.sent;
            resourceIds.forEach(function (resourceId, index) {
              manifests[resourceId] = manifestArray[index];
            });
            return _context.abrupt("return", manifests);

          case 10:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));
  return _fetchResourceManifests.apply(this, arguments);
}

;

function getLanguages(_x2) {
  return _getLanguages.apply(this, arguments);
}

function _getLanguages() {
  _getLanguages = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee2(_ref3) {
    var username, resourceIds, languageIds, languages;
    return regeneratorRuntime.wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            username = _ref3.username, resourceIds = _ref3.resourceIds;
            _context2.next = 3;
            return getLanguageIds({
              username: username,
              resourceIds: resourceIds
            });

          case 3:
            languageIds = _context2.sent;
            languages = languageIds.map(function (languageId) {
              return (0, _languages.getLanguage)({
                languageId: languageId
              });
            }).filter(function (language) {
              return !!language;
            });
            languages.sort(function (a, b) {
              return a.languageId > b.languageId ? 1 : b.languageId > a.languageId ? -1 : 0;
            });
            return _context2.abrupt("return", languages);

          case 7:
          case "end":
            return _context2.stop();
        }
      }
    }, _callee2);
  }));
  return _getLanguages.apply(this, arguments);
}

;

function getLanguageIds(_x3) {
  return _getLanguageIds.apply(this, arguments);
} // /repos/search?q=ulb&uid=4598&limit=50&exclusive=true


function _getLanguageIds() {
  _getLanguageIds = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee3(_ref4) {
    var username, resourceIds, languageIds, promises, languageIdsArray, _languageIds;

    return regeneratorRuntime.wrap(function _callee3$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            username = _ref4.username, resourceIds = _ref4.resourceIds;
            languageIds = [];
            promises = resourceIds.map(function (resourceId) {
              return getLanguageIdsByResource({
                username: username,
                resourceId: resourceId
              });
            });
            _context3.next = 5;
            return Promise.all(promises);

          case 5:
            languageIdsArray = _context3.sent;
            _languageIds = languageIdsArray.flat();

            _languageIds.forEach(function (languageId) {
              var languageAdded = languageIds.includes(languageId);
              if (!languageAdded) languageIds.push(languageId);
            });

            return _context3.abrupt("return", languageIds);

          case 9:
          case "end":
            return _context3.stop();
        }
      }
    }, _callee3);
  }));
  return _getLanguageIds.apply(this, arguments);
}

function getLanguageIdsByResource(_x4) {
  return _getLanguageIdsByResource.apply(this, arguments);
}

function _getLanguageIdsByResource() {
  _getLanguageIdsByResource = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee4(_ref5) {
    var username, resourceId, languageIds, uid, params, url, repos;
    return regeneratorRuntime.wrap(function _callee4$(_context4) {
      while (1) {
        switch (_context4.prev = _context4.next) {
          case 0:
            username = _ref5.username, resourceId = _ref5.resourceId;
            languageIds = [];
            _context4.next = 4;
            return (0, _.getUID)({
              username: username
            });

          case 4:
            uid = _context4.sent;
            params = {
              q: resourceId,
              uid: uid,
              limit: 50,
              exclusive: true
            };
            url = _path.default.join(apiPath, "repos/search");
            _context4.next = 9;
            return (0, _.get)({
              url: url,
              params: params
            });

          case 9:
            repos = _context4.sent;

            if (repos && repos.data) {
              languageIds = repos.data.map(function (repo) {
                return repo.name.split('_')[0];
              });
            }

            return _context4.abrupt("return", languageIds);

          case 12:
          case "end":
            return _context4.stop();
        }
      }
    }, _callee4);
  }));
  return _getLanguageIdsByResource.apply(this, arguments);
}

;

function fetchManifest(_x5) {
  return _fetchManifest.apply(this, arguments);
}

function _fetchManifest() {
  _fetchManifest = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee5(_ref6) {
    var username, repository, yaml, json;
    return regeneratorRuntime.wrap(function _callee5$(_context5) {
      while (1) {
        switch (_context5.prev = _context5.next) {
          case 0:
            username = _ref6.username, repository = _ref6.repository;
            _context5.next = 3;
            return (0, _.getFile)({
              username: username,
              repository: repository,
              path: 'manifest.yaml'
            });

          case 3:
            yaml = _context5.sent;
            json = yaml ? _jsYamlParser.default.safeLoad(yaml) : null;
            return _context5.abrupt("return", json);

          case 6:
          case "end":
            return _context5.stop();
        }
      }
    }, _callee5);
  }));
  return _fetchManifest.apply(this, arguments);
}

;

function fetchRepositoriesZipFiles(_x6) {
  return _fetchRepositoriesZipFiles.apply(this, arguments);
}

function _fetchRepositoriesZipFiles() {
  _fetchRepositoriesZipFiles = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee6(_ref7) {
    var username, languageId, branch, repositories, promises, zipArray;
    return regeneratorRuntime.wrap(function _callee6$(_context6) {
      while (1) {
        switch (_context6.prev = _context6.next) {
          case 0:
            username = _ref7.username, languageId = _ref7.languageId, branch = _ref7.branch;
            repositories = resourceRepositories({
              languageId: languageId
            });
            promises = Object.values(repositories).map(function (repository) {
              return (0, _.fetchRepositoryZipFile)({
                username: username,
                repository: repository,
                branch: branch
              });
            });
            _context6.next = 5;
            return Promise.all(promises);

          case 5:
            zipArray = _context6.sent;
            return _context6.abrupt("return", zipArray);

          case 7:
          case "end":
            return _context6.stop();
        }
      }
    }, _callee6);
  }));
  return _fetchRepositoriesZipFiles.apply(this, arguments);
}

;
