// gitea-react-toolkit/dist/core/gitea-api/fetch/zip.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getFileFromZip = exports.fetchRepositoryZipFile = void 0;

var _path = _interopRequireDefault(require("path"));

var _jszip = _interopRequireDefault(require("jszip"));

var _localforage = _interopRequireDefault(require("localforage"));

var _repos = require("../repos");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

var zipStore = _localforage.default.createInstance({
  driver: [_localforage.default.INDEXEDDB],
  name: 'zip-store'
}); // https://bg.door43.org/unfoldingWord/en_ult/archive/master.zip


var fetchRepositoryZipFile = /*#__PURE__*/function () {
  var _ref2 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee(_ref) {
    var username, repository, branch, options, repoExists, uri, response, zipArrayBuffer;
    return regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            username = _ref.username, repository = _ref.repository, branch = _ref.branch, options = _ref.options;
            _context.next = 3;
            return (0, _repos.repositoryExists)({
              username: username,
              repository: repository
            });

          case 3:
            repoExists = _context.sent;

            if (repoExists) {
              _context.next = 6;
              break;
            }

            return _context.abrupt("return", null);

          case 6:
            uri = zipUri({
              username: username,
              repository: repository,
              branch: branch
            });
            _context.next = 9;
            return fetch(uri);

          case 9:
            response = _context.sent;

            if (!(response.status === 200 || response.status === 0)) {
              _context.next = 19;
              break;
            }

            _context.next = 13;
            return response.arrayBuffer();

          case 13:
            zipArrayBuffer = _context.sent;
            _context.next = 16;
            return zipStore.setItem(uri, zipArrayBuffer);

          case 16:
            return _context.abrupt("return", true);

          case 19:
            return _context.abrupt("return", false);

          case 20:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));

  return function fetchRepositoryZipFile(_x) {
    return _ref2.apply(this, arguments);
  };
}();

exports.fetchRepositoryZipFile = fetchRepositoryZipFile;

var getFileFromZip = /*#__PURE__*/function () {
  var _ref4 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee2(_ref3) {
    var username, repository, path, branch, file, uri, zipBlob, zip, zipPath;
    return regeneratorRuntime.wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            username = _ref3.username, repository = _ref3.repository, path = _ref3.path, branch = _ref3.branch;
            uri = zipUri({
              username: username,
              repository: repository,
              branch: branch
            });
            _context2.next = 4;
            return zipStore.getItem(uri);

          case 4:
            zipBlob = _context2.sent;
            _context2.prev = 5;

            if (!zipBlob) {
              _context2.next = 14;
              break;
            }

            _context2.next = 9;
            return _jszip.default.loadAsync(zipBlob);

          case 9:
            zip = _context2.sent;
            zipPath = _path.default.join(repository.toLowerCase(), path);
            _context2.next = 13;
            return zip.file(zipPath).async("string");

          case 13:
            file = _context2.sent;

          case 14:
            _context2.next = 19;
            break;

          case 16:
            _context2.prev = 16;
            _context2.t0 = _context2["catch"](5);
            file = null;

          case 19:
            return _context2.abrupt("return", file);

          case 20:
          case "end":
            return _context2.stop();
        }
      }
    }, _callee2, null, [[5, 16]]);
  }));

  return function getFileFromZip(_x2) {
    return _ref4.apply(this, arguments);
  };
}();

exports.getFileFromZip = getFileFromZip;

var zipUri = function zipUri(_ref5) {
  var username = _ref5.username,
      repository = _ref5.repository,
      _ref5$branch = _ref5.branch,
      branch = _ref5$branch === void 0 ? 'master' : _ref5$branch;

  var zipPath = _path.default.join(username, repository, 'archive', "".concat(branch, ".zip"));

  var zipUri = baseURL + zipPath;
  return zipUri;
};
