// gitea-react-toolkit/dist/components/repository/helpers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getFileFromRepositoryZip = exports.removeRepositoryZip = exports.storeRepositoryZip = exports.deleteRepository = exports.saveRepository = exports.readRepository = exports.createRepository = exports.repositoryForks = exports.forkRepository = void 0;

var _ = require("../..");

var _2 = require(".");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

var forkRepository = /*#__PURE__*/function () {
  var _ref2 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee(_ref) {
    var repository, config, username, name, response;
    return regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            repository = _ref.repository, config = _ref.config;
            username = repository.owner.username, name = repository.name;
            _context.next = 4;
            return (0, _.createFork)({
              owner: username,
              repo: name,
              config: config
            });

          case 4:
            response = _context.sent;
            return _context.abrupt("return", response);

          case 6:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));

  return function forkRepository(_x) {
    return _ref2.apply(this, arguments);
  };
}();

exports.forkRepository = forkRepository;

var repositoryForks = /*#__PURE__*/function () {
  var _ref4 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee2(_ref3) {
    var repository, config, username, name, response;
    return regeneratorRuntime.wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            repository = _ref3.repository, config = _ref3.config;
            username = repository.owner.username, name = repository.name;
            _context2.next = 4;
            return (0, _.readForks)({
              owner: username,
              repo: name,
              config: config
            });

          case 4:
            response = _context2.sent;
            return _context2.abrupt("return", response);

          case 6:
          case "end":
            return _context2.stop();
        }
      }
    }, _callee2);
  }));

  return function repositoryForks(_x2) {
    return _ref4.apply(this, arguments);
  };
}();

exports.repositoryForks = repositoryForks;

var createRepository = /*#__PURE__*/function () {
  var _ref6 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee3(_ref5) {
    var repo, settings, config, _settings, response;

    return regeneratorRuntime.wrap(function _callee3$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            repo = _ref5.repo, settings = _ref5.settings, config = _ref5.config;
            _settings = _objectSpread({
              name: repo,
              description: "".concat(repo || settings.name, " created via API."),
              private: false
            }, settings);
            _context3.next = 4;
            return (0, _.createRepo)({
              repo: repo,
              settings: _settings,
              config: config
            });

          case 4:
            response = _context3.sent;
            return _context3.abrupt("return", response);

          case 6:
          case "end":
            return _context3.stop();
        }
      }
    }, _callee3);
  }));

  return function createRepository(_x3) {
    return _ref6.apply(this, arguments);
  };
}();

exports.createRepository = createRepository;

var readRepository = /*#__PURE__*/function () {
  var _ref8 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee4(_ref7) {
    var owner, repo, config, response;
    return regeneratorRuntime.wrap(function _callee4$(_context4) {
      while (1) {
        switch (_context4.prev = _context4.next) {
          case 0:
            owner = _ref7.owner, repo = _ref7.repo, config = _ref7.config;
            _context4.next = 3;
            return (0, _.readRepo)({
              owner: owner,
              repo: repo,
              config: config
            });

          case 3:
            response = _context4.sent;
            return _context4.abrupt("return", response);

          case 5:
          case "end":
            return _context4.stop();
        }
      }
    }, _callee4);
  }));

  return function readRepository(_x4) {
    return _ref8.apply(this, arguments);
  };
}();

exports.readRepository = readRepository;

var saveRepository = /*#__PURE__*/function () {
  var _ref10 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee5(_ref9) {
    var repository, settings, config, username, name, response;
    return regeneratorRuntime.wrap(function _callee5$(_context5) {
      while (1) {
        switch (_context5.prev = _context5.next) {
          case 0:
            repository = _ref9.repository, settings = _ref9.settings, config = _ref9.config;
            username = repository.owner.username, name = repository.name;
            _context5.next = 4;
            return (0, _.updateRepo)({
              owner: username,
              repo: name,
              settings: settings,
              config: config
            });

          case 4:
            response = _context5.sent;
            return _context5.abrupt("return", response);

          case 6:
          case "end":
            return _context5.stop();
        }
      }
    }, _callee5);
  }));

  return function saveRepository(_x5) {
    return _ref10.apply(this, arguments);
  };
}();

exports.saveRepository = saveRepository;

var deleteRepository = /*#__PURE__*/function () {
  var _ref12 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee6(_ref11) {
    var repository, settings, config, username, name, response;
    return regeneratorRuntime.wrap(function _callee6$(_context6) {
      while (1) {
        switch (_context6.prev = _context6.next) {
          case 0:
            repository = _ref11.repository, settings = _ref11.settings, config = _ref11.config;
            username = repository.owner.username, name = repository.name;
            _context6.next = 4;
            return (0, _.deleteRepo)({
              owner: username,
              repo: name,
              settings: settings,
              config: config
            });

          case 4:
            response = _context6.sent;
            return _context6.abrupt("return", response);

          case 6:
          case "end":
            return _context6.stop();
        }
      }
    }, _callee6);
  }));

  return function deleteRepository(_x6) {
    return _ref12.apply(this, arguments);
  };
}();

exports.deleteRepository = deleteRepository;

var storeRepositoryZip = /*#__PURE__*/function () {
  var _ref14 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee7(_ref13) {
    var repository, config, username, name, branch, response;
    return regeneratorRuntime.wrap(function _callee7$(_context7) {
      while (1) {
        switch (_context7.prev = _context7.next) {
          case 0:
            repository = _ref13.repository, config = _ref13.config;
            username = repository.owner.username, name = repository.name, branch = repository.branch;
            _context7.next = 4;
            return (0, _2.storeZipFromRepo)({
              owner: username,
              repo: name,
              branch: branch,
              server: config.server
            });

          case 4:
            response = _context7.sent;
            return _context7.abrupt("return", response);

          case 6:
          case "end":
            return _context7.stop();
        }
      }
    }, _callee7);
  }));

  return function storeRepositoryZip(_x7) {
    return _ref14.apply(this, arguments);
  };
}();

exports.storeRepositoryZip = storeRepositoryZip;

var removeRepositoryZip = /*#__PURE__*/function () {
  var _ref16 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee8(_ref15) {
    var repository, config, username, name, branch, response;
    return regeneratorRuntime.wrap(function _callee8$(_context8) {
      while (1) {
        switch (_context8.prev = _context8.next) {
          case 0:
            repository = _ref15.repository, config = _ref15.config;
            username = repository.owner.username, name = repository.name, branch = repository.branch;
            _context8.next = 4;
            return (0, _2.storeZipFromRepo)({
              owner: username,
              repo: name,
              branch: branch,
              server: config.server
            });

          case 4:
            response = _context8.sent;
            return _context8.abrupt("return", response);

          case 6:
          case "end":
            return _context8.stop();
        }
      }
    }, _callee8);
  }));

  return function removeRepositoryZip(_x8) {
    return _ref16.apply(this, arguments);
  };
}();

exports.removeRepositoryZip = removeRepositoryZip;

var getFileFromRepositoryZip = /*#__PURE__*/function () {
  var _ref18 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee9(_ref17) {
    var repository, filepath, config, username, name, branch, file;
    return regeneratorRuntime.wrap(function _callee9$(_context9) {
      while (1) {
        switch (_context9.prev = _context9.next) {
          case 0:
            repository = _ref17.repository, filepath = _ref17.filepath, config = _ref17.config;
            username = repository.owner.username, name = repository.name, branch = repository.branch;
            _context9.next = 4;
            return (0, _2.getFileFromRepoZip)({
              owner: username,
              repo: name,
              branch: branch,
              filepath: filepath,
              config: config
            });

          case 4:
            file = _context9.sent;
            return _context9.abrupt("return", file);

          case 6:
          case "end":
            return _context9.stop();
        }
      }
    }, _callee9);
  }));

  return function getFileFromRepositoryZip(_x9) {
    return _ref18.apply(this, arguments);
  };
}();

exports.getFileFromRepositoryZip = getFileFromRepositoryZip;
