// markdown-translatable/dist/components/block-editable/useStyles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useStyles = void 0;

var _styles = require("@material-ui/core/styles");

var useStyles = (0, _styles.makeStyles)(function (theme) {
  return {
    root: {
      height: '100%',
      width: '100%'
    },
    html: {
      'height': '100%',
      'width': 'calc(100% - 1em)',
      'display': 'grid',
      'padding': '0 0.5em',
      'lineHeight': '1.4',
      '&:focus-visible': {
        outline: '#38ADDF auto 1px'
      }
    },
    markdown: {
      'height': '100%',
      'width': 'calc(100% - 1em)',
      'display': 'grid',
      'whiteSpace': 'pre-wrap',
      'paddingBlockStart': '1em',
      'paddingBlockEnd': '1em',
      'fontFamily': 'unset',
      'padding': '0 0.5em',
      'lineHeight': '1.4',
      'fontSize': 'unset',
      '&:focus-visible': {
        outline: '#38ADDF auto 1px'
      }
    },
    pre: {
      margin: 0,
      fontFamily: 'unset',
      fontSize: 'unset',
      whiteSpace: 'normal'
    }
  };
});
exports.useStyles = useStyles;
