// gitea-react-toolkit/dist/components/authentication/useAuthentication.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _deepFreeze = _interopRequireDefault(require("deep-freeze"));

var _core = require("../../core");

var _ = require(".");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function useAuthentication(_ref) {
  var messages = _ref.messages,
      _authentication = _ref.authentication,
      onAuthentication = _ref.onAuthentication,
      config = _ref.config,
      loadAuthentication = _ref.loadAuthentication,
      saveAuthentication = _ref.saveAuthentication,
      onError = _ref.onError;

  var authentication = _authentication && (0, _deepFreeze.default)(_authentication);

  var _useState = (0, _react.useState)(),
      _useState2 = _slicedToArray(_useState, 2),
      error = _useState2[0],
      setError = _useState2[1];

  if (!messages) {
    messages = _core.defaultErrorMessages;
  }

  var logout = (0, _react.useCallback)(function (_auth) {
    saveAuthentication && saveAuthentication();
  }, [saveAuthentication]);
  /*
  const update = useCallback(async (_auth) => {
    if (_auth && _auth.remember && saveAuthentication) {
      await saveAuthentication(_auth);
    }
    return onAuthentication && onAuthentication(_auth);
  }, [onAuthentication]);
  */

  var update = (0, _react.useCallback)(function (_auth) {
    if (_auth) {
      if (saveAuthentication) {
        if (_auth.remember) {
          saveAuthentication(_auth);
        } else {
          saveAuthentication();
        }
      }
    }

    if (onAuthentication) {
      onAuthentication(_auth);
    }
  }, [onAuthentication, saveAuthentication]);
  (0, _react.useEffect)(function () {
    if (!authentication && loadAuthentication) {
      loadAuthentication().then(function (_authentication) {
        if (_authentication) {
          update(_authentication);
        }
      });
    }
  }, [authentication, loadAuthentication, update]);
  var onSubmitLogin = (0, _react.useCallback)( /*#__PURE__*/function () {
    var _ref3 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee(_ref2) {
      var username, password, remember, _authentication2, user, token, friendlyError;

      return regeneratorRuntime.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              username = _ref2.username, password = _ref2.password, remember = _ref2.remember;
              _context.prev = 1;
              _context.next = 4;
              return (0, _core.authenticate)({
                username: username,
                password: password,
                config: config
              });

            case 4:
              _authentication2 = _context.sent;
              _authentication2.remember = remember;

              if (_authentication2) {
                user = _authentication2.user, token = _authentication2.token;

                if (user && token) {
                  setError();
                  update(_authentication2);
                } else {
                  if (!user) {
                    setError(messages.usernameError);
                  } else if (!token) {
                    setError(messages.passwordError);
                  }
                }
              } else {
                console.log('authentication failed?', _authentication2);
              }

              _context.next = 15;
              break;

            case 9:
              _context.prev = 9;
              _context.t0 = _context["catch"](1);
              console.log('Authentication error:', _context.t0);
              friendlyError = (0, _core.parseError)(_context.t0);
              setError(friendlyError.errorMessage);
              onError && onError(friendlyError);

            case 15:
            case "end":
              return _context.stop();
          }
        }
      }, _callee, null, [[1, 9]]);
    }));

    return function (_x) {
      return _ref3.apply(this, arguments);
    };
  }(), [config, logout, update, messages.genericError, messages.networkError, messages.passwordError, messages.serverError, messages.usernameError]);
  var onSubmit = (0, _react.useCallback)( /*#__PURE__*/function () {
    var _ref5 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee2(_ref4) {
      var username, password, remember;
      return regeneratorRuntime.wrap(function _callee2$(_context2) {
        while (1) {
          switch (_context2.prev = _context2.next) {
            case 0:
              username = _ref4.username, password = _ref4.password, remember = _ref4.remember;

              if (!authentication) {
                _context2.next = 6;
                break;
              }

              logout();
              update();
              _context2.next = 8;
              break;

            case 6:
              _context2.next = 8;
              return onSubmitLogin({
                username: username,
                password: password,
                remember: remember
              });

            case 8:
            case "end":
              return _context2.stop();
          }
        }
      }, _callee2);
    }));

    return function (_x2) {
      return _ref5.apply(this, arguments);
    };
  }(), [authentication, config, logout, update, onSubmitLogin]);
  var component = (0, _react.useMemo)(function () {
    return config && /*#__PURE__*/_react.default.createElement(_.LoginForm, {
      config: config,
      authentication: authentication,
      actionText: messages.actionText,
      errorText: error,
      onSubmit: onSubmit
    });
  }, [authentication, config, error, messages.actionText, onSubmit]);

  var _config = authentication && authentication.config || config;

  var response = {
    state: authentication,
    actions: {
      update: update,
      logout: logout,
      onLoginFormSubmit: onSubmit,
      onLoginFormSubmitLogin: onSubmitLogin
    },
    component: component,
    config: _config,
    messages: messages
  };
  return response;
}

;
useAuthentication.propTypes = {
  /** Pass a previously returned authentication object to bypass login. */
  authentication: _propTypes.default.shape({
    user: _propTypes.default.object.isRequired,
    token: _propTypes.default.object.isRequired,
    config: _propTypes.default.object.isRequired,
    remember: _propTypes.default.bool
  }),

  /** Callback function to propogate the user/token used for API Authentication. */
  onAuthentication: _propTypes.default.func,

  /** Configuration to pass through to the Authentication component. */

  /** Override the default text and errors. Must override all or none. */
  messages: _propTypes.default.shape({
    actionText: _propTypes.default.string.isRequired,
    genericError: _propTypes.default.string.isRequired,
    usernameError: _propTypes.default.string.isRequired,
    passwordError: _propTypes.default.string.isRequired
  }),
  config: _propTypes.default.shape({
    /** The Gitea server to use when authenticating. */
    server: _propTypes.default.string.isRequired,

    /** The id of the token to create/retrieve that is used for the app. */
    tokenid: _propTypes.default.string.isRequired
  }),

  /** Callback function to persist authentication. */
  saveAuthentication: _propTypes.default.func,

  /** Callback function to retrieve persisted authentication. */
  loadAuthentication: _propTypes.default.func,

  /** Callback function on error. */
  onError: _propTypes.default.func
};
var _default = useAuthentication;
exports.default = _default;
