// gitea-react-toolkit/dist/core/languages/languages.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getLanguageName = exports.getLanguage = void 0;

var _langnames = _interopRequireDefault(require("./langnames.json"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var getLanguage = function getLanguage(_ref) {
  var languageId = _ref.languageId;
  var language;

  var langname = _langnames.default.filter(function (object) {
    return object.lc === languageId;
  })[0];

  if (langname) {
    language = {
      id: langname.pk,
      languageName: langname.ang,
      region: langname.lr,
      gateway: langname.gw,
      country: langname.hc,
      localized: langname.ln,
      languageId: langname.lc,
      direction: langname.ld,
      aliases: langname.alt,
      countries: langname.cc
    };
  }

  return language;
};

exports.getLanguage = getLanguage;

var getLanguageName = function getLanguageName(_ref2) {
  var languageId = _ref2.languageId;
  var language = getLanguage({
    languageId: languageId
  });
  var languageName = language ? language.localized : null;
  return languageName;
};

exports.getLanguageName = getLanguageName;
