// gitea-react-toolkit/dist/core/gitea-api/users/tokens.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ensureToken = exports.deleteToken = exports.createToken = exports.getTokens = void 0;

var _path = _interopRequireDefault(require("path"));

var _ = require("..");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _createForOfIteratorHelper(o, allowArrayLike) { var it; if (typeof Symbol === "undefined" || o[Symbol.iterator] == null) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = o[Symbol.iterator](); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it.return != null) it.return(); } finally { if (didErr) throw err; } } }; }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

// requires config.headers with authorization
var getTokens = /*#__PURE__*/function () {
  var _ref2 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee(_ref) {
    var username, config, tokens, url;
    return regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            username = _ref.username, config = _ref.config;
            url = _path.default.join(_.apiPath, 'users', username, 'tokens');
            _context.prev = 2;
            _context.next = 5;
            return (0, _.get)({
              url: url,
              config: config
            });

          case 5:
            tokens = _context.sent;
            _context.next = 11;
            break;

          case 8:
            _context.prev = 8;
            _context.t0 = _context["catch"](2);
            tokens = null;

          case 11:
            return _context.abrupt("return", tokens);

          case 12:
          case "end":
            return _context.stop();
        }
      }
    }, _callee, null, [[2, 8]]);
  }));

  return function getTokens(_x) {
    return _ref2.apply(this, arguments);
  };
}();

exports.getTokens = getTokens;

// requires config.headers with authorization
var createToken = /*#__PURE__*/function () {
  var _ref4 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee2(_ref3) {
    var username, config, token, url, payload;
    return regeneratorRuntime.wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            username = _ref3.username, config = _ref3.config;
            url = _path.default.join(_.apiPath, 'users', username, 'tokens');
            payload = {
              'name': config.tokenid
            };
            _context2.prev = 3;
            _context2.next = 6;
            return (0, _.post)({
              url: url,
              payload: payload,
              config: config
            });

          case 6:
            token = _context2.sent;
            _context2.next = 12;
            break;

          case 9:
            _context2.prev = 9;
            _context2.t0 = _context2["catch"](3);
            token = undefined;

          case 12:
            return _context2.abrupt("return", token);

          case 13:
          case "end":
            return _context2.stop();
        }
      }
    }, _callee2, null, [[3, 9]]);
  }));

  return function createToken(_x2) {
    return _ref4.apply(this, arguments);
  };
}();

exports.createToken = createToken;

// requires config.headers with authorization
var deleteToken = /*#__PURE__*/function () {
  var _ref6 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee3(_ref5) {
    var username, token, config, success, id, url;
    return regeneratorRuntime.wrap(function _callee3$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            username = _ref5.username, token = _ref5.token, config = _ref5.config;
            id = token.id.toString();
            url = _path.default.join(_.apiPath, 'users', username, 'tokens', id);
            _context3.prev = 3;
            _context3.next = 6;
            return (0, _.del)({
              url: url,
              config: config,
              payload: {}
            });

          case 6:
            success = true;
            _context3.next = 12;
            break;

          case 9:
            _context3.prev = 9;
            _context3.t0 = _context3["catch"](3);
            success = false;

          case 12:
            return _context3.abrupt("return", success);

          case 13:
          case "end":
            return _context3.stop();
        }
      }
    }, _callee3, null, [[3, 9]]);
  }));

  return function deleteToken(_x3) {
    return _ref6.apply(this, arguments);
  };
}(); // requires config.headers with authorization


exports.deleteToken = deleteToken;

var ensureToken = /*#__PURE__*/function () {
  var _ref8 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee4(_ref7) {
    var username, config, tokens, tokenMatches, _iterator, _step, _token2, token;

    return regeneratorRuntime.wrap(function _callee4$(_context4) {
      while (1) {
        switch (_context4.prev = _context4.next) {
          case 0:
            username = _ref7.username, config = _ref7.config;
            _context4.next = 3;
            return getTokens({
              username: username,
              config: config
            });

          case 3:
            tokens = _context4.sent;

            if (!tokens) {
              _context4.next = 24;
              break;
            }

            tokenMatches = tokens.filter(function (_token) {
              return _token.name === config.tokenid;
            });

            if (!(tokenMatches && tokenMatches.length)) {
              _context4.next = 24;
              break;
            }

            _iterator = _createForOfIteratorHelper(tokenMatches);
            _context4.prev = 8;

            _iterator.s();

          case 10:
            if ((_step = _iterator.n()).done) {
              _context4.next = 16;
              break;
            }

            _token2 = _step.value;
            _context4.next = 14;
            return deleteToken({
              username: username,
              token: _token2,
              config: config
            });

          case 14:
            _context4.next = 10;
            break;

          case 16:
            _context4.next = 21;
            break;

          case 18:
            _context4.prev = 18;
            _context4.t0 = _context4["catch"](8);

            _iterator.e(_context4.t0);

          case 21:
            _context4.prev = 21;

            _iterator.f();

            return _context4.finish(21);

          case 24:
            _context4.next = 26;
            return createToken({
              username: username,
              config: config
            });

          case 26:
            token = _context4.sent;
            return _context4.abrupt("return", token);

          case 28:
          case "end":
            return _context4.stop();
        }
      }
    }, _callee4, null, [[8, 18, 21, 24]]);
  }));

  return function ensureToken(_x4) {
    return _ref8.apply(this, arguments);
  };
}();

exports.ensureToken = ensureToken;
