// gitea-react-toolkit/dist/components/authentication/Authentication.context.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AuthenticationContextProvider = AuthenticationContextProvider;
exports.AuthenticationContext = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _ = require(".");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var AuthenticationContext = _react.default.createContext();

exports.AuthenticationContext = AuthenticationContext;

function AuthenticationContextProvider(_ref) {
  var _config = _ref.config,
      messages = _ref.messages,
      authentication = _ref.authentication,
      onAuthentication = _ref.onAuthentication,
      loadAuthentication = _ref.loadAuthentication,
      saveAuthentication = _ref.saveAuthentication,
      children = _ref.children,
      onError = _ref.onError;

  var _useAuthentication = (0, _.useAuthentication)({
    authentication: authentication,
    onAuthentication: onAuthentication,
    config: _config,
    messages: messages,
    loadAuthentication: loadAuthentication,
    saveAuthentication: saveAuthentication,
    onError: onError
  }),
      state = _useAuthentication.state,
      actions = _useAuthentication.actions,
      component = _useAuthentication.component,
      config = _useAuthentication.config;

  var context = {
    state: state,
    actions: actions,
    component: component,
    config: config
  };
  return /*#__PURE__*/_react.default.createElement(AuthenticationContext.Provider, {
    value: context
  }, children);
}

;
AuthenticationContextProvider.propTypes = {
  /** Pass a previously returned authentication object to bypass login. */
  authentication: _propTypes.default.shape({
    user: _propTypes.default.object.isRequired,
    token: _propTypes.default.object.isRequired,
    config: _propTypes.default.object.isRequired,
    remember: _propTypes.default.bool
  }),

  /** Callback function to propogate the user/token used for API Authentication. */
  onAuthentication: _propTypes.default.func,

  /** Configuration to pass through to the Authentication component. */

  /** Override the default text and errors. Must override all or none. */
  messages: _propTypes.default.shape({
    actionText: _propTypes.default.string.isRequired,
    genericError: _propTypes.default.string.isRequired,
    usernameError: _propTypes.default.string.isRequired,
    passwordError: _propTypes.default.string.isRequired
  }),
  config: _propTypes.default.shape({
    /** The Gitea server to use when authenticating. */
    server: _propTypes.default.string.isRequired,

    /** The id of the token to create/retrieve that is used for the app. */
    tokenid: _propTypes.default.string.isRequired
  }),

  /** Callback function to persist authentication. */
  saveAuthentication: _propTypes.default.func,

  /** Callback function to retrieve persisted authentication. */
  loadAuthentication: _propTypes.default.func,

  /** optional callback for error */
  onError: _propTypes.default.func
};
