// gitea-react-toolkit/dist/components/authentication/LoginForm.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LoginFormComponent = void 0;

var _react = _interopRequireWildcard(require("react"));

var _useDeepCompareEffect = _interopRequireDefault(require("use-deep-compare-effect"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _core = require("@material-ui/core");

var _icons = require("@material-ui/icons");

var _useStyles = require("./useStyles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function LoginForm(_ref) {
  var config = _ref.config,
      authentication = _ref.authentication,
      actionText = _ref.actionText,
      errorText = _ref.errorText,
      onSubmit = _ref.onSubmit;
  var classes = (0, _useStyles.useStyles)();

  var _authentication = authentication || {
    user: {}
  };

  var _username = _authentication.user.username;
  var _remember = _authentication.remember;

  var _useState = (0, _react.useState)({
    username: _username,
    remember: _remember
  }),
      _useState2 = _slicedToArray(_useState, 2),
      formData = _useState2[0],
      setFormData = _useState2[1];

  var username = formData.username,
      password = formData.password,
      remember = formData.remember;
  (0, _useDeepCompareEffect.default)(function () {
    if (_authentication) {
      var _formData = _objectSpread({}, formData);

      if (_authentication.user.username) {
        _formData.username = _authentication.user.username;
      }

      if (_authentication.remember) {
        _formData.remember = _authentication.remember;
      }

      setFormData(_formData);
    } else {
      var _formData2 = _objectSpread({}, formData);

      _formData2.password = '';
      setFormData(_formData2);
    }
  }, [_authentication]);
  var user;
  if (authentication) user = authentication.user;

  var updateFormData = function updateFormData(event) {
    var _event$target = event.target,
        type = _event$target.type,
        name = _event$target.name,
        value = _event$target.value,
        checked = _event$target.checked;

    var _formData = _objectSpread({}, formData);

    if (type === 'checkbox') _formData[value] = checked;else _formData[name] = value;
    setFormData(_formData);
  };

  var footer = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null);

  if (config && config.server) {
    footer = /*#__PURE__*/_react.default.createElement("div", {
      className: classes.footer_container
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: classes.footer_column
    }, /*#__PURE__*/_react.default.createElement(_core.Typography, {
      variant: "caption"
    }, "Need an account?\xA0"), /*#__PURE__*/_react.default.createElement(_core.Typography, {
      color: "primary",
      variant: "caption",
      component: "a",
      target: "_blank",
      href: "".concat(config.server, "/user/sign_up")
    }, "Register now.")), /*#__PURE__*/_react.default.createElement("div", {
      className: classes.footer_column
    }, /*#__PURE__*/_react.default.createElement(_core.Typography, {
      color: "primary",
      variant: "caption",
      component: "a",
      target: "_blank",
      href: "".concat(config.server, "/user/forgot_password")
    }, "Forgot password?")));
  }

  return /*#__PURE__*/_react.default.createElement("div", {
    className: classes.root
  }, /*#__PURE__*/_react.default.createElement(_core.Avatar, {
    className: classes.avatar,
    src: user && user.avatar_url ? user.avatar_url : null
  }, /*#__PURE__*/_react.default.createElement(_icons.LockOutlined, null)), /*#__PURE__*/_react.default.createElement(_core.Typography, {
    component: "h1",
    variant: "h5"
  }, user ? user.full_name : actionText), errorText ? /*#__PURE__*/_react.default.createElement(_core.Typography, {
    "data-test": "login-error-text",
    component: "p",
    style: {
      color: 'red'
    }
  }, errorText) : /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null), /*#__PURE__*/_react.default.createElement("form", {
    className: classes.form
  }, /*#__PURE__*/_react.default.createElement(_core.TextField, {
    "data-test": "username-input",
    name: "username",
    type: "text",
    label: "Username",
    required: true,
    variant: "outlined",
    margin: "normal",
    fullWidth: true,
    autoComplete: "username",
    disabled: !!user,
    defaultValue: username,
    onChange: updateFormData
  }), /*#__PURE__*/_react.default.createElement(_core.TextField, {
    "data-test": "password-input",
    name: "password",
    type: "password",
    label: "Password",
    required: true,
    variant: "outlined",
    margin: "normal",
    fullWidth: true,
    autoComplete: "current-password",
    disabled: !!user,
    onChange: updateFormData
  }), /*#__PURE__*/_react.default.createElement(_core.FormControlLabel, {
    "data-test": "remember-checkbox",
    label: "Keep me logged in",
    control: /*#__PURE__*/_react.default.createElement(_core.Checkbox, {
      color: "primary",
      value: "remember",
      disabled: !!user,
      checked: remember,
      id: 'remember-' + Math.random(),
      onChange: updateFormData
    })
  }), /*#__PURE__*/_react.default.createElement(_core.Button, {
    "data-test": user ? 'logout-button' : 'submit-button',
    type: "button",
    fullWidth: true,
    variant: "contained",
    color: "primary",
    className: classes.submit,
    onClick: function onClick() {
      return onSubmit(formData);
    }
  }, user ? 'Logout' : actionText), footer));
}

;
LoginForm.propTypes = {
  /** Configuration to use for sign up/forgot password flow */
  config: _propTypes.default.shape({
    server: _propTypes.default.string.isRequired
  }),

  /** Callback function to propogate the username and password entered. */
  onSubmit: _propTypes.default.func.isRequired,

  /** The text to describe the action of logging in. */
  actionText: _propTypes.default.string,

  /** The text to describe the error when Authentication fails. */
  errorText: _propTypes.default.string,

  /** The authenticated user object */
  authentication: _propTypes.default.shape({
    user: _propTypes.default.object
  })
};
LoginForm.defaultProps = {
  actionText: 'Login'
};
var LoginFormComponent = LoginForm;
exports.LoginFormComponent = LoginFormComponent;
