// datatable-translatable/dist/components/datatable/helpers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getColumns = getColumns;
exports.getData = getData;

var _react = _interopRequireDefault(require("react"));

var _helpers = require("../column-filter/helpers");

var _cell = require("../cell");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && Symbol.iterator in Object(iter)) return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function getColumns(_ref) {
  var columnNames = _ref.columnNames,
      columnsFilter = _ref.columnsFilter,
      columnsFilterOptions = _ref.columnsFilterOptions,
      columnsShow = _ref.columnsShow,
      delimiters = _ref.delimiters,
      rowHeader = _ref.rowHeader,
      generateRowId = _ref.generateRowId,
      cellEdit = _ref.cellEdit,
      preview = _ref.preview;
  var columns = columnNames.map(function (_name) {
    var name = _name === null || _name === void 0 ? void 0 : _name.trim();
    var offset = rowHeader ? 1 : 0;
    var filterOptions;

    if (columnsFilter.includes(name)) {
      filterOptions = {
        logic: function logic(value, filters) {
          return (0, _helpers.filterLogic)({
            value: value,
            filters: filters,
            delimiters: delimiters
          });
        },
        display: function display(filterList, onChange, filterIndex, column) {
          return (0, _helpers.filterDisplay)({
            filterList: filterList,
            onChange: onChange,
            column: column,
            offset: offset,
            columnsFilterOptions: columnsFilterOptions,
            filterIndex: filterIndex
          });
        }
      };
    }

    ;
    return {
      name: name,
      searchable: true,
      options: {
        display: columnsShow.includes(name),
        filter: columnsFilter.includes(name),
        filterType: columnsFilter.includes(name) ? 'custom' : undefined,
        filterOptions: filterOptions,
        customBodyRender: function customBodyRender(value, tableMeta) {
          var _tableMeta$tableState = tableMeta.tableState,
              tableState = _tableMeta$tableState === void 0 ? {} : _tableMeta$tableState;

          var _ref2 = tableState || {},
              rowsPerPage = _ref2.rowsPerPage,
              page = _ref2.page;

          var cellProps = {
            generateRowId: generateRowId,
            value: value,
            tableMeta: tableMeta,
            onEdit: cellEdit,
            delimiters: delimiters,
            rowsPerPage: rowsPerPage,
            page: page,
            preview: preview
          };
          return /*#__PURE__*/_react["default"].createElement(_cell.Cell, cellProps);
        },
        customFilterListOptions: {
          render: function render(value) {
            return "".concat(name, " - ").concat(value);
          }
        }
      }
    };
  });

  if (rowHeader) {
    var headerColumn = {
      name: 'rowHeader',
      options: {
        filter: false,
        customBodyRender: function customBodyRender(_, tableMeta) {
          var cellProps = {
            generateRowId: generateRowId,
            tableMeta: tableMeta,
            delimiters: delimiters,
            rowHeader: rowHeader
          };
          return /*#__PURE__*/_react["default"].createElement(_cell.HeaderCell, cellProps);
        }
      }
    };
    columns.unshift(headerColumn);
  }

  return columns;
}

function getData(_ref3) {
  var data = _ref3.data,
      columnNames = _ref3.columnNames,
      rowHeader = _ref3.rowHeader;

  var _data = _toConsumableArray(data);

  if (columnNames && data && rowHeader) {
    _data = data.map(function (row) {
      return ['rowHeader'].concat(_toConsumableArray(row));
    });
  }

  return _data;
}
