// datatable-translatable/dist/components/datatable/muiTheme.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getMuiTheme = void 0;

var _styles = require("@material-ui/core/styles");

var getMuiTheme = (0, _styles.createMuiTheme)({
  typography: {
    useNextVariants: true
  },
  overrides: {
    MuiIconButton: {
      root: {
        padding: '8px'
      }
    },
    MuiTableCell: {
      root: {
        'padding': '0',
        'textAlign': 'unset',
        'display': 'block',
        'borderBottom': 'none',
        '&:last-child': {
          paddingRight: 0
        },
        // '&:first-child': { top:'0 !important' },
        '&:nth-child(1)': {
          position: 'sticky',
          top: '48px',
          background: 'white'
        }
      },
      body: {
        fontSize: '1em'
      }
    },
    MuiTableRow: {
      root: {
        height: 'unset',
        padding: '0',
        display: 'block'
      }
    },
    MUIDataTable: {
      root: {},
      responsiveScroll: {
        maxHeight: 'unset',
        overflowX: 'unset',
        overflowY: 'unset'
      },
      paper: {
        boxShadow: '0px'
      }
    },
    MuiToolbar: {
      root: {
        top: 0,
        position: 'sticky',
        background: 'white',
        zIndex: '100',
        minHeight: '48px !important',
        height: '48px !important'
      }
    },
    MUIDataTableHeadRow: {
      root: {
        display: 'none'
      }
    },
    MuiTable: {
      root: {
        position: 'sticky',
        bottom: 0,
        background: 'white'
      }
    },
    MuiTableFooter: {
      root: {
        borderTop: '1px solid #ccc'
      }
    }
  }
});
exports.getMuiTheme = getMuiTheme;
